{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "10d6e92e-e382-4ad3-bf20-d3ffd6ed5c4f",
   "metadata": {},
   "source": [
    "# ⚠️⚠️ **WARNING: DO NOT EXECUTE AS-IS** ⚠️⚠️\n",
    "\n",
    "<h3 style=\"color:red;\">⚠️⚠️ Running the following cells will overwrite previously saved results ⚠️⚠️</h3>\n",
    "\n",
    "- These operations will **replace** files on disk.\n",
    "- If unsure, contact Mikaela.\n",
    "- If something was run by mistake, please contact Mikaela, most files can be recovered for ~1 week from over-writing.\n",
    "\n",
    "➡️ **Recommended:** Run only after verifying that cache/output directories are altered.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877901af-1216-4a08-925b-cb464100ffdd",
   "metadata": {},
   "source": [
    "# Notebook Summary\n",
    "This notebook is for loading data from source through filtering stages.\n",
    "\n",
    "Starts with raw data (species cluster data from GTDB, metadata from GTDB, and metadata from NCBI).\n",
    "\n",
    "Then filtering stage1 based on minimum clade size and list of NCBI warnings and version status.\n",
    "\n",
    "Next is stage2 on sourmash for pre-ANI filtering.\n",
    "\n",
    "Next is stage3, code is both for prepping the fastANI runs and processing based on choosen ANI threshold.\n",
    "\n",
    "Then cells to prep for mOTUpan (input file and genome pathlist creation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "796a25fb-0e10-4cfb-8192-ed8500c28e68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Notebook Settings & Imports\n",
    "#   Note: if you can't load the dataframes please ensure pandas is >v2.0.0, known backcompatibility issues exists\n",
    "#-------------------------\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import importlib\n",
    "import pickle\n",
    "import os\n",
    "import re\n",
    "\n",
    "# Custom functions\n",
    "import pangenome_utils as pg\n",
    "\n",
    "# Jupyter settings\n",
    "#default=last_expr #(https://ipython.readthedocs.io/en/stable/config/options/terminal.html#configtrait-InteractiveShell.ast_node_interactivity)\n",
    "from IPython.display import display\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "#InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7ce9b70d-7ab6-4b52-b5ae-30309aec17d7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define filters for Stage1 (clade size and NCBI)\n",
    "#-------------------------\n",
    "# Minimum clade size is 2, filter out NCBI suppressed OR any of listed warnings\n",
    "\n",
    "filter_min_clade_size = 2\n",
    "filter_out_ncbi_version_status = ['suppressed']\n",
    "filter_out_ncbi_warnings = ['fragmented assembly','genome length too large','genome length too small','hybrid','low quality sequence','misassembled','partial','annotation fails completeness check','low gene count','RefSeq annotation failed','many frameshifted proteins']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0be9d454-5b86-4b0c-be09-ae598d6ca0f9",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 01 -- Load in raw source datasets\n",
    "\n",
    "Raw unfiltered data (one cell per primary dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f1bd9ce-bc45-4b31-98c6-53521d5cc926",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.5 s, sys: 94.1 ms, total: 1.6 s\n",
      "Wall time: 1.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------    \n",
    "# Load species cluster data from GTDB (this is the base dataset)\n",
    "#     Output: df_gtdb_species_data_unfiltered\n",
    "#     Dependencies: None\n",
    "#-------------------------\n",
    "# Last runtime: 1.8 s\n",
    "\n",
    "# Load from file\n",
    "sp_cluster_file_path = \"/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/GTDB_r214/sp_clusters_r214.tsv\"\n",
    "df_gtdb_species_data_unfiltered = pd.read_csv(sp_cluster_file_path, sep='\\t')\n",
    "\n",
    "# Add grouping by ANI\n",
    "#-------------------------\n",
    "ASSIGN_ANIGROUP = True\n",
    "def assign_group(row):\n",
    "    if row['Mean intra-species ANI'] == 100:\n",
    "        return 'Group1: 100'\n",
    "    elif 99.9 >= row['Mean intra-species ANI'] > 99.5:\n",
    "        return 'Group2: 99.9-99.5'\n",
    "    elif 99.5 >= row['Mean intra-species ANI'] > 99.0:\n",
    "        return 'Group3: 99.5-99.0'\n",
    "    elif 99.0 >= row['Mean intra-species ANI'] > 98.5:\n",
    "        return 'Group4: 99.0-98.5'\n",
    "    elif 98.5 >= row['Mean intra-species ANI'] > 98.0:\n",
    "        return 'Group5: 98.5-98.0'\n",
    "    elif 98.0 >= row['Mean intra-species ANI'] > 97.0:\n",
    "        return 'Group6: 98.0-97.0'\n",
    "    else:\n",
    "        return 'Group7: LT 97.0'\n",
    "if ASSIGN_ANIGROUP: df_gtdb_species_data_unfiltered['Group'] = df_gtdb_species_data_unfiltered.apply(assign_group, axis=1)\n",
    "\n",
    "# Break up the species data into taxonomic columns\n",
    "temp_df = df_gtdb_species_data_unfiltered['GTDB taxonomy'].str.split(';', expand=True)\n",
    "temp_df.columns = ['domain', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "df_gtdb_species_data_unfiltered = pd.concat([df_gtdb_species_data_unfiltered, temp_df], axis=1)\n",
    "\n",
    "# Add clade_name\n",
    "df_gtdb_species_data_unfiltered['clade_name']=df_gtdb_species_data_unfiltered['GTDB species']+\"--\"+df_gtdb_species_data_unfiltered['Representative genome']\n",
    "df_gtdb_species_data_unfiltered['clade_name'].replace(' ', '_', regex=True, inplace=True)\n",
    "\n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_species_data_unfiltered.pkl', 'wb') as file: pickle.dump(df_gtdb_species_data_unfiltered, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db9d03f6-45e6-436f-9197-c51797a7ff8c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<timed exec>:16: DtypeWarning: Columns (34,35,36) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 18.3 s, sys: 3.07 s, total: 21.3 s\n",
      "Wall time: 22.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Load in all four NCBI report dataframes and concatenate into one\n",
    "#     Output: df_ncbi_reportdata_unfiltered\n",
    "#     Dependencies: None\n",
    "#-------------------------\n",
    "# Last runtime: 22.9 s\n",
    "\n",
    "# Read in all four dataframes\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_genbank_historical.txt'\n",
    "df_ncbi_report_data_genbank_historical = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_genbank.txt'\n",
    "df_ncbi_report_data_genbank = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_refseq_historical.txt'\n",
    "df_ncbi_report_data_refseq_historical = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_refseq.txt'\n",
    "df_ncbi_report_data_refseq = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "\n",
    "# Concatenate\n",
    "df_ncbi_reportdata_unfiltered = pd.concat([df_ncbi_report_data_genbank_historical, df_ncbi_report_data_genbank, df_ncbi_report_data_refseq_historical, df_ncbi_report_data_refseq], axis=0, ignore_index=True)\n",
    "for d in [df_ncbi_report_data_genbank_historical, df_ncbi_report_data_genbank, df_ncbi_report_data_refseq_historical, df_ncbi_report_data_refseq]: del d\n",
    "\n",
    "# Ensure the \"Warnings\" column is properly parsed as a list of strings\n",
    "#df_ncbi_metadata['Warnings'] = df_ncbi_metadata['Warnings'].apply(eval)\n",
    "df_ncbi_reportdata_unfiltered['excluded_from_refseq'] = df_ncbi_reportdata_unfiltered['excluded_from_refseq'].str.split('; ').apply(lambda x: x if isinstance(x, list) else [x])\n",
    "\n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_ncbi_reportdata_unfiltered.pkl', 'wb') as file: pickle.dump(df_ncbi_reportdata_unfiltered, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8b37213f-bfbf-49c6-808d-5cc991361d3f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<timed exec>:16: DtypeWarning: Columns (34,35,36) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 19.3 s, sys: 2.95 s, total: 22.3 s\n",
      "Wall time: 22.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Load in all four NCBI report dataframes and concatenate into one\n",
    "#     Output: df_ncbi_reportdata_unfiltered\n",
    "#     Dependencies: None\n",
    "#-------------------------\n",
    "# Last runtime: 22.9 s\n",
    "\n",
    "# Read in all four dataframes\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_genbank_historical.txt'\n",
    "df_ncbi_report_data_genbank_historical = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_genbank.txt'\n",
    "df_ncbi_report_data_genbank = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_refseq_historical.txt'\n",
    "df_ncbi_report_data_refseq_historical = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "tsv_file = '/global/cfs/cdirs/kbase/ke_prototype/mcashman/data/NCBI_datasets/NCBI_reports/assembly_summary_refseq.txt'\n",
    "df_ncbi_report_data_refseq = pd.read_csv(tsv_file, delimiter=\"\\t\", header=1)\n",
    "\n",
    "# Concatenate\n",
    "df_ncbi_reportdata_unfiltered = pd.concat([df_ncbi_report_data_genbank_historical, df_ncbi_report_data_genbank, df_ncbi_report_data_refseq_historical, df_ncbi_report_data_refseq], axis=0, ignore_index=True)\n",
    "for d in [df_ncbi_report_data_genbank_historical, df_ncbi_report_data_genbank, df_ncbi_report_data_refseq_historical, df_ncbi_report_data_refseq]: del d\n",
    "\n",
    "# Ensure the \"Warnings\" column is properly parsed as a list of strings\n",
    "#df_ncbi_metadata['Warnings'] = df_ncbi_metadata['Warnings'].apply(eval)\n",
    "df_ncbi_reportdata_unfiltered['excluded_from_refseq'] = df_ncbi_reportdata_unfiltered['excluded_from_refseq'].str.split('; ').apply(lambda x: x if isinstance(x, list) else [x])\n",
    "\n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_ncbi_reportdata_unfiltered.pkl', 'wb') as file: pickle.dump(df_ncbi_reportdata_unfiltered, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4ea1b5-80ca-4b9d-a62c-ec7c475894c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------------------    \n",
    "# Readers for in dataframe created above (toggle as desired)\n",
    "#-------------------------\n",
    "print(\"Dataframes available:\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_species_data_unfiltered.pkl', 'rb') as file: df_gtdb_species_data_unfiltered = pickle.load(file); print(\"  df_gtdb_species_data_unfiltered\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_ncbi_reportdata_unfiltered.pkl', 'rb') as file: df_ncbi_reportdata_unfiltered = pickle.load(file); print(\"  df_ncbi_reportdata_unfiltered\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_metadata_unfiltered.pkl', 'rb') as file: df_gtdb_metadata_unfiltered = pickle.load(file); print(\"  df_gtdb_metadata_unfiltered\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f497c970-3a94-497c-9408-986d3380181f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 02 -- Perform Stage1 filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5bf14a8-8e64-4c0d-871c-3c22e38f9221",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup1: Define base filtering dataframe (list of all genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "275a0c59-b2cd-4bb4-acc7-14707675d2b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Create a master filter dataframe to track filters and act as key for applying to dataframes\n",
    "#     Output: df_master_filter_keys\n",
    "#     Dependencies: df_gtdb_species_data_unfiltered\n",
    "#-------------------------\n",
    "# Last runtime: 694 ms\n",
    "\n",
    "# Read in source data\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_species_data_unfiltered.pkl', 'rb') as file: df_gtdb_species_data_unfiltered = pickle.load(file)\n",
    "\n",
    "# Explode species data to get all genomes\n",
    "df_gtdb_species_data_unfiltered['Clustered genomes'] = df_gtdb_species_data_unfiltered['Clustered genomes'].str.split(',')\n",
    "df_master_filter_keys = df_gtdb_species_data_unfiltered.explode('Clustered genomes')\n",
    "\n",
    "# Reorder columns as needed\n",
    "df_master_filter_keys = df_master_filter_keys[['clade_name', 'Representative genome', 'GTDB species', 'No. clustered genomes', 'GTDB taxonomy', 'Clustered genomes', 'domain', 'phylum', 'class', 'order', 'family', 'genus']]\n",
    "\n",
    "# Sanity check\n",
    "if ~(len(df_master_filter_keys) == df_gtdb_species_data_unfiltered['No. clustered genomes'].sum()):\n",
    "    print(\"Lengths do not match, something went wrong\")\n",
    "    exit(1)\n",
    "    \n",
    "# Reorder and rename columns   \n",
    "df_master_filter_keys.insert(0, 'genome_ID', df_master_filter_keys.pop('Clustered genomes'))\n",
    "df_master_filter_keys = df_master_filter_keys.rename(columns={'No. clustered genomes': 'No. clustered genomes in species clade'})\n",
    "\n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'wb') as file: pickle.dump(df_master_filter_keys, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0530d81-5841-40bc-abd3-d0f82b258a98",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup2: Add first stage of filters to master_filter_keys dataframe (overwrites base df)\n",
    "\n",
    "- FilterA  : clade size\n",
    "- filterA1 : clade size < 2\n",
    "- FilterB  : NCBI warnings\n",
    "- FilterB1 : NCBI warnings defined above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "711c66ab-ba6e-4425-a18b-a4ea614dd39e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number to filter out w.r.t. clade size: 56567\n",
      "Number additional to filter out w.r.t. NCBI flags: 14839\n",
      "CPU times: user 6.92 s, sys: 1.28 s, total: 8.2 s\n",
      "Wall time: 8.36 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Create filter columns in master filter dataframe (1 is fits filter, 0 is not)\n",
    "#     Output: df_master_filter_keys (more columns)\n",
    "#     Dependencies: df_master_filter_keys, df_ncbi_reportdata_unfiltered\n",
    "#-------------------------\n",
    "# Last runtime: 8.36 s\n",
    "\n",
    "# Read in any source data needed\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_ncbi_reportdata_unfiltered.pkl', 'rb') as file: df_ncbi_reportdata_unfiltered = pickle.load(file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file)\n",
    "\n",
    "#-- Filter based on clade size (FilterA)\n",
    "df_master_filter_keys['FilterA1'] = np.where(df_master_filter_keys['No. clustered genomes in species clade'] < filter_min_clade_size, 1, 0)\n",
    "\n",
    "#-- Filter based on NCBI warnings (FilterB)\n",
    "# Get list of accessions to filter out\n",
    "accessions_to_filter_out_ncbi = df_ncbi_reportdata_unfiltered[\n",
    "    ((df_ncbi_reportdata_unfiltered['version_status'].apply(lambda x: x in filter_out_ncbi_version_status))\n",
    "    | (df_ncbi_reportdata_unfiltered['excluded_from_refseq'].apply(lambda x: any(item in filter_out_ncbi_warnings for item in x))))\n",
    "]['assembly_accession']\n",
    "# Cross with dataframe\n",
    "df_master_filter_keys['FilterB1'] = np.where(df_master_filter_keys['genome_ID'].str[3:].isin(accessions_to_filter_out_ncbi.to_list()), 1, 0)\n",
    "\n",
    "# Print summary\n",
    "print(f\"Number to filter out w.r.t. clade size: {df_master_filter_keys['FilterA1'].sum()}\")\n",
    "print(f\"Number additional to filter out w.r.t. NCBI flags: {df_master_filter_keys[df_master_filter_keys['FilterA1']==0]['FilterB1'].sum()}\")\n",
    "df_master_filter_keys\n",
    "\n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'wb') as file: pickle.dump(df_master_filter_keys, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b9c077-1376-46b6-a87d-9311da067df3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stage1 Part1 : Filter the species clusters first as any cluster remaining with < the min clade size must be added to the filter_out list\n",
    "\n",
    "- Adds in FilterC_A1B1 which is anything new to filter out (e.g. FilterC) once Filters A1 and B1 are in place (e.g. clades might be left with only 1 genome)\n",
    "- Thus updates master_keys df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "2639b8ff-b91f-47f7-a194-f9905bde7566",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.86 s, sys: 160 ms, total: 2.02 s\n",
      "Wall time: 2.03 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Filter down raw datasets.  Have to start with gtdb_species as can cause more genomes to be filtered out when clade drops below threshold\n",
    "#\n",
    "#     Output: df_gtdb_species_data_filteredStage1, df_master_filter_keys\n",
    "#     Dependencies: df_gtdb_species_data_unfiltered, df_master_filter_keys\n",
    "#-------------------------\n",
    "# Last runtime: 2.03 s\n",
    "\n",
    "# Load in any dependencies\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_species_data_unfiltered.pkl', 'rb') as file: df_gtdb_species_data_unfiltered = pickle.load(file)\n",
    "\n",
    "# Get list of genomes IDs to drop\n",
    "accessions_to_filter_out = df_master_filter_keys[(df_master_filter_keys['FilterA1']==1) | (df_master_filter_keys['FilterB1']==1)]['genome_ID'].to_list()\n",
    "accessions_to_keep       = df_master_filter_keys[(df_master_filter_keys['FilterA1']==0) & (df_master_filter_keys['FilterB1']==0)]['genome_ID'].to_list()\n",
    "accessions_to_filter_out_noprefix = [a[3:] for a in accessions_to_filter_out]\n",
    "accessions_to_keep_noprefix       = [a[3:] for a in accessions_to_keep]\n",
    "\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_gtdb_metadata_unfiltered\n",
    "#     Output: df_gtdb_metadata_filteredStage1\n",
    "#-------------------------\n",
    "df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_unfiltered.copy()\n",
    "accessions_to_filter_out_set = set(accessions_to_filter_out)\n",
    "\n",
    "# Step 1: Convert \"Clustered genomes\" column to a list\n",
    "df_gtdb_species_data_filteredStage1[\"Clustered genomes\"] = df_gtdb_species_data_filteredStage1[\"Clustered genomes\"].str.split(\",\")\n",
    "\n",
    "# Step 2: Filter the list of genomes in each clade\n",
    "# Create a log of any genomes to be removed in new column\n",
    "df_gtdb_species_data_filteredStage1[\"Clustered genomes removed\"] = (\n",
    "    df_gtdb_species_data_filteredStage1[\"Clustered genomes\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "# Update the \"Clustered genomes\" column after applying the filters\n",
    "df_gtdb_species_data_filteredStage1[\"Clustered genomes\"] = (\n",
    "    df_gtdb_species_data_filteredStage1[\"Clustered genomes\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: ~x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "\n",
    "# Step 3: Calculate the length of the updated \"Clustered genomes\" list and create a new column, rename accordingly\n",
    "df_gtdb_species_data_filteredStage1[\"No. clustered genomes Filtered Stage1\"] = df_gtdb_species_data_filteredStage1[\"Clustered genomes\"].apply(lambda x: len(x) if isinstance(x, list) else 0)\n",
    "df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_filteredStage1.rename(columns={'No. clustered genomes': 'No. clustered genomes Unfiltered', 'Clustered genomes': 'Clustered genomes retained'})\n",
    "\n",
    "# Step 4: Filter again w.r.t. size\n",
    "new_accessions_to_filter_out = df_gtdb_species_data_filteredStage1[df_gtdb_species_data_filteredStage1[\"No. clustered genomes Filtered Stage1\"] < filter_min_clade_size][\"Clustered genomes retained\"]\n",
    "filtered_list = [x for x in new_accessions_to_filter_out.to_list() if str(x) != 'nan']\n",
    "new_accessions_to_filter_out = [item for sublist in filtered_list for item in sublist]\n",
    "#df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_filteredStage1[df_gtdb_species_data_filteredStage1['Filtered No. clustered genomes'] >= filter_min_clade_size]\n",
    "\n",
    "# Step 5: Finish filtering the dataframe (drop any < min size) and save\n",
    "df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_filteredStage1[df_gtdb_species_data_filteredStage1['No. clustered genomes Filtered Stage1'] >= filter_min_clade_size]\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_gtdb_species_data_filteredStage1.pkl', 'wb') as file: pickle.dump(df_gtdb_species_data_filteredStage1, file)\n",
    "\n",
    "#-------------------------\n",
    "# Update df_master_filter_keys with any additional genomes that got dropped\n",
    "#     Output: df_master_filter_keys\n",
    "#-------------------------\n",
    "df_master_filter_keys['FilterC_A1B1'] = np.where(df_master_filter_keys['genome_ID'].isin(new_accessions_to_filter_out), 1, 0)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'wb') as file: pickle.dump(df_master_filter_keys, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82a9d679-5dbb-4665-b59b-8d8438e39fd8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Stage1 Part2 : With any added filters in place, filter down reamining raw data sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f59958e0-ae5b-4dca-a3bc-021ce7451625",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.1 s, sys: 1.97 s, total: 13.1 s\n",
      "Wall time: 13.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------\n",
    "# Filter down remaining raw datasets\n",
    "#\n",
    "#     Output: df_ncbi_reportdata_filteredStage1, df_gtdb_metadata_filteredStage1\n",
    "#     Dependencies: df_master_filter_keys, df_ncbi_reportdata_unfiltered, df_gtdb_metadata_unfiltered\n",
    "#-------------------------\n",
    "# Last runtime: 13.6 s\n",
    "\n",
    "# Load in any dependencies\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_ncbi_reportdata_unfiltered.pkl', 'rb') as file: df_ncbi_reportdata_unfiltered = pickle.load(file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/unfiltered/df_gtdb_metadata_unfiltered.pkl', 'rb') as file: df_gtdb_metadata_unfiltered = pickle.load(file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file)\n",
    "\n",
    "# Get list of genomes IDs to drop\n",
    "accessions_to_filter_out = df_master_filter_keys[(df_master_filter_keys['FilterA1']==1) | (df_master_filter_keys['FilterB1']==1) | (df_master_filter_keys['FilterC_A1B1']==1)]['genome_ID'].to_list()\n",
    "accessions_to_keep       = df_master_filter_keys[(df_master_filter_keys['FilterA1']==0) & (df_master_filter_keys['FilterB1']==0) & (df_master_filter_keys['FilterC_A1B1']==0)]['genome_ID'].to_list()\n",
    "accessions_to_filter_out_noprefix = [a[3:] for a in accessions_to_filter_out]\n",
    "accessions_to_keep_noprefix       = [a[3:] for a in accessions_to_keep]\n",
    "\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_ncbi_reportdata_unfiltered\n",
    "#     Output: df_ncbi_reportdata_filteredStage1\n",
    "#-------------------------\n",
    "df_ncbi_reportdata_filteredStage1 = df_ncbi_reportdata_unfiltered[df_ncbi_reportdata_unfiltered['assembly_accession'].isin(accessions_to_keep_noprefix)]\n",
    "if not(len(df_ncbi_reportdata_filteredStage1)==len(accessions_to_keep_noprefix)): print(\"----!!! WARNING---- NCBI report data missing some genomes to retain\") \n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_ncbi_reportdata_filteredStage1.pkl', 'wb') as file: pickle.dump(df_ncbi_reportdata_filteredStage1, file)\n",
    "\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_gtdb_metadata_unfiltered\n",
    "#     Output: df_gtdb_metadata_filteredStage1\n",
    "#-------------------------\n",
    "df_gtdb_metadata_filteredStage1 = df_gtdb_metadata_unfiltered[df_gtdb_metadata_unfiltered['accession'].isin(accessions_to_keep)]\n",
    "if not(len(df_gtdb_metadata_filteredStage1)==len(accessions_to_keep)): print(\"----!!! WARNING---- GTDB metadata missing some genomes to retain\") \n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_gtdb_metadata_filteredStage1.pkl', 'wb') as file: pickle.dump(df_gtdb_metadata_filteredStage1, file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de02f81-f24b-4e20-be03-5834ee763343",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 03 -- Perform Stage2 filtering - Sourmash filter\n",
    "Add in an additional Filter marker for the sourmash filtering (currently just on Ecoli)\n",
    "\n",
    "Modifies:\n",
    "- Master key dataframe (df_master_filter_keys)\n",
    "- 3 other dataframes: df_gtdb_species_data_filteredStage2, df_ncbi_reportdata_filteredStage2, df_gtdb_metadata_filteredStage2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2bbf31f6-c4a8-4705-801a-2c88810a0f07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of additional genomes to filter out 18003\n",
      "Total number of clades changed 1\n",
      "Done\n",
      "CPU times: user 5.07 s, sys: 1.49 s, total: 6.57 s\n",
      "Wall time: 6.96 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------    \n",
    "# Fetch list of genomes to be removed\n",
    "#-------------------------\n",
    "# Last runtime: 6.96 s\n",
    "\n",
    "# Open master keys\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file)\n",
    "# Open list of genomes to be removed (already calculated with a sourmash pipeline probably on the GPU nodes)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage2/list_of_removed_genomes_s__Escherichia_coli--RS_GCF_003697165.2_0.9992.txt', 'rb') as file: list_of_removed_genomes_s__Escherichia_coli = [line.strip().decode('utf-8') for line in file]\n",
    "\n",
    "# Add in a column to the master filtering dataframe\n",
    "df_master_filter_keys['FilterD1'] = np.where(df_master_filter_keys['genome_ID'].str[3:].isin(list_of_removed_genomes_s__Escherichia_coli), 1, 0)\n",
    "# Get list of genomes IDs to drop\n",
    "accessions_to_filter_out = df_master_filter_keys[(df_master_filter_keys['FilterD1']==1)]['genome_ID'].to_list()\n",
    "print(f\"Total number of additional genomes to filter out {len(accessions_to_filter_out)}\")\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_gtdb_metadata_filteredStage1\n",
    "#     Output: df_gtdb_metadata_filteredStage2\n",
    "#-------------------------\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_gtdb_species_data_filteredStage1.pkl', 'rb') as file: df_gtdb_species_data_filteredStage1 = pickle.load(file)\n",
    "df_gtdb_species_data_filteredStage2 = df_gtdb_species_data_filteredStage1.copy()\n",
    "accessions_to_filter_out_set = set(accessions_to_filter_out)\n",
    "\n",
    "# Step 1: Filter the list of genomes in each clade\n",
    "# Create a log of any genomes to be removed in new column\n",
    "df_gtdb_species_data_filteredStage2[\"Clustered genomes removed Stage2\"] = (\n",
    "    df_gtdb_species_data_filteredStage2[\"Clustered genomes retained\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "# Update the \"Clustered genomes\" column after applying the filters\n",
    "df_gtdb_species_data_filteredStage2[\"Clustered genomes retained\"] = (\n",
    "    df_gtdb_species_data_filteredStage2[\"Clustered genomes retained\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: ~x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "\n",
    "# Step 2: Calculate the length of the updated \"Clustered genomes\" list and create a new column, rename accordingly\n",
    "df_gtdb_species_data_filteredStage2[\"No. clustered genomes Filtered Stage2\"] = df_gtdb_species_data_filteredStage2[\"Clustered genomes retained\"].apply(lambda x: len(x) if isinstance(x, list) else 0)\n",
    "df_gtdb_species_data_filteredStage2 = df_gtdb_species_data_filteredStage2.rename(columns={'No. clustered genomes': 'No. clustered genomes Filter Stage1', 'Clustered genomes retained': 'Clustered genomes retained Stage2'})\n",
    "\n",
    "# Step 3: Filter again w.r.t. size\n",
    "new_accessions_to_filter_out = df_gtdb_species_data_filteredStage2[df_gtdb_species_data_filteredStage2[\"No. clustered genomes Filtered Stage2\"] < filter_min_clade_size][\"Clustered genomes retained Stage2\"]\n",
    "filtered_list = [x for x in new_accessions_to_filter_out.to_list() if str(x) != 'nan']\n",
    "new_accessions_to_filter_out = [item for sublist in filtered_list for item in sublist]\n",
    "#df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_filteredStage1[df_gtdb_species_data_filteredStage1['Filtered No. clustered genomes'] >= filter_min_clade_size]\n",
    "\n",
    "# Stop here if any clades dropped below threshold\n",
    "if len(df_gtdb_species_data_filteredStage2[df_gtdb_species_data_filteredStage2['No. clustered genomes Filtered Stage2']<filter_min_clade_size])>0:\n",
    "    print(f\"----WARNING---- Some clades dropped in size due to Stage 2 filter, please alter code\")\n",
    "    exit(1)\n",
    "\n",
    "# Step 4: Finish filtering the dataframe (drop any < min size) and save\n",
    "#df_gtdb_species_data_filteredStage2 = df_gtdb_species_data_filteredStage2[df_gtdb_species_data_filteredStage2['No. clustered genomes Filtered Stage2'] >= filter_min_clade_size]\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_species_data_filteredStage2.pkl', 'wb') as file: pickle.dump(df_gtdb_species_data_filteredStage2, file)\n",
    "\n",
    "# Print out many rows changed as a sanity check\n",
    "row_change_count = len(df_gtdb_species_data_filteredStage2[~(df_gtdb_species_data_filteredStage2['No. clustered genomes Filtered Stage1']==df_gtdb_species_data_filteredStage2['No. clustered genomes Filtered Stage2'])])\n",
    "print(f\"Total number of clades changed {row_change_count}\")\n",
    "\n",
    "\n",
    "\n",
    "#-------------------------\n",
    "# Filter down remaining raw datasets\n",
    "#\n",
    "#     Output: df_ncbi_reportdata_filteredStage2, df_gtdb_metadata_filteredStage2\n",
    "#     Dependencies: df_master_filter_keys, df_ncbi_reportdata_filteredStage1, df_gtdb_metadata_filteredStage1\n",
    "#-------------------------\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_ncbi_reportdata_filteredStage1.pkl', 'rb') as file: df_ncbi_reportdata_filteredStage1 = pickle.load(file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage1/df_gtdb_metadata_filteredStage1.pkl', 'rb') as file: df_gtdb_metadata_filteredStage1 = pickle.load(file)\n",
    "\n",
    "# Get list of genomes IDs to drop\n",
    "accessions_to_filter_out = df_master_filter_keys[(df_master_filter_keys['FilterD1']==1)]['genome_ID'].to_list()\n",
    "#accessions_to_keep       = df_master_filter_keys[(df_master_filter_keys['FilterD1']==0)]['genome_ID'].to_list()\n",
    "accessions_to_filter_out_noprefix = [a[3:] for a in accessions_to_filter_out]\n",
    "#accessions_to_keep_noprefix       = [a[3:] for a in accessions_to_keep]\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_ncbi_reportdata_filteredStage1\n",
    "#     Output: df_ncbi_reportdata_filteredStage2\n",
    "#-------------------------\n",
    "df_ncbi_reportdata_filteredStage2 = df_ncbi_reportdata_filteredStage1[~df_ncbi_reportdata_filteredStage1['assembly_accession'].isin(accessions_to_filter_out_noprefix)]\n",
    "#if not(len(df_ncbi_reportdata_filteredStage2)==len(accessions_to_keep_noprefix)): print(\"----!!! WARNING---- NCBI report data missing some genomes to retain\") \n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_ncbi_reportdata_filteredStage2.pkl', 'wb') as file: pickle.dump(df_ncbi_reportdata_filteredStage2, file)\n",
    "\n",
    "\n",
    "#-------------------------\n",
    "# Perform the filtering on df_gtdb_metadata_filteredStage1\n",
    "#     Output: df_gtdb_metadata_filteredStage2\n",
    "#-------------------------\n",
    "df_gtdb_metadata_filteredStage2 = df_gtdb_metadata_filteredStage1[~df_gtdb_metadata_filteredStage1['accession'].isin(accessions_to_filter_out)]\n",
    "#if not(len(df_gtdb_metadata_filteredStage2)==len(accessions_to_keep)): print(\"----!!! WARNING---- GTDB metadata missing some genomes to retain\") \n",
    "# Save dataframe to file (keeping data types)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_metadata_filteredStage2.pkl', 'wb') as file: pickle.dump(df_gtdb_metadata_filteredStage2, file)\n",
    "\n",
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cad7e29-79ed-4ce1-abb9-937888d2a89b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# /End/ Dataframe readers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6ede75a-f4fd-4888-bc6f-f2c1f81fbeb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------------------    \n",
    "# Readers for any dataframe created above (toggle as desired)\n",
    "#-------------------------\n",
    "print(\"Dataframes available:\")\n",
    "#-- Master Keys\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file); print(\"  df_master_filter_keys\")\n",
    "#-- Source Data\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_species_data_filteredStage2.pkl', 'rb') as file: df_gtdb_species_data_filteredStage2 = pickle.load(file); print(\"  df_gtdb_species_data_filteredStage2\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_ncbi_reportdata_filteredStage2.pkl', 'rb') as file: df_ncbi_reportdata_filteredStage2 = pickle.load(file); print(\"  df_ncbi_reportdata_filteredStage2\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_metadata_filteredStage2.pkl', 'rb') as file: df_gtdb_metadata_filteredStage2 = pickle.load(file); print(\"  df_gtdb_metadata_filteredStage2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "688017a7-3607-4eab-8f13-5766485060e8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# -- STEP04 -- Run fastANI\n",
    "\n",
    "\n",
    "1. Generate genome fna pathlists on NERSC with `dataframes/filtering_stage3/generate_genome_pathlists_fna.py`\n",
    "    - Input: speies_data dataframe for the list of genomes\n",
    "    - Output: genomes_pathlist for each clade in user-defined output directory (see filtered_stage3)\n",
    "\n",
    "2. Run fastANI in `/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/pairwise_ANI/species_level_runs`\n",
    "    - Previously run on both NERSC and KBase GPU (58 largest plus 2 stragglers)\n",
    "    - Now all result files back on NERSC in that dir in RESULTS/MERGED_OUT\n",
    "\n",
    "3. Process results (two stages)\n",
    "    - STEP01 process the results per_clade `/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP01_process_fastANI` \n",
    "    - STEP02 process the per_clade results per_genomepair `/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "dcdc2a6b-c140-4aac-bd71-1957ee289b58",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Dataframes available (fastANI results Per Clade):\n",
      "df_fastANI_byClade_LT500\n",
      "\n",
      ">>> Dataframes available (fastANI results Per Genome Pair):\n",
      "df_fastANI_byGenomePair_LT500\n",
      "df_fastANI_byGenomePair_combinedGTE500sm\n",
      "df_fastANI_byGenomePair_SE\n",
      "df_fastANI_byGenomePair_KP\n",
      "df_fastANI_byGenomePair_SA\n",
      "df_fastANI_byGenomePair_EC\n",
      "\n",
      ">>> Complete Dataframes available (fastANI results Per Genome Pair):\n",
      "df_fastANI_byGenomePair_combinedALL\n"
     ]
    }
   ],
   "source": [
    "#-------------------------    \n",
    "# Readers for any dataframe created above (toggle as desired)\n",
    "#-------------------------\n",
    "print(\">>> Dataframes available (fastANI results Per Clade):\")\n",
    "for x in ['df_fastANI_byClade_LT500']: print(x)\n",
    "#-- fastANI results Per Clade\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP01_process_fastANI/df_fastANI_byClade_LT500.pkl', 'rb') as file: df_fastANI_byClade_LT500 = pickle.load(file); print(\"  df_fastANI_byClade_LT500\")\n",
    "\n",
    "print(\"\\n>>> Dataframes available (fastANI results Per Genome Pair):\")\n",
    "for x in ['df_fastANI_byGenomePair_LT500','df_fastANI_byGenomePair_combinedGTE500sm','df_fastANI_byGenomePair_SE','df_fastANI_byGenomePair_KP','df_fastANI_byGenomePair_SA','df_fastANI_byGenomePair_EC']: print(x)\n",
    "#-- fastANI results Per Genome Pair\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_LT500.pkl', 'rb') as file: df_fastANI_byGenomePair_LT500 = pickle.load(file); print(\"  df_fastANI_byGenomePair_LT500\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_combinedGTE500sm.pkl', 'rb') as file: df_fastANI_byGenomePair_combinedGTE500sm = pickle.load(file); print(\"  df_fastANI_byGenomePair_combinedGTE500sm\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_SE.pkl', 'rb') as file: df_fastANI_byGenomePair_SE = pickle.load(file); print(\"  df_fastANI_byGenomePair_SE\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_KP.pkl', 'rb') as file: df_fastANI_byGenomePair_KP = pickle.load(file); print(\"  df_fastANI_byGenomePair_KP\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_SA.pkl', 'rb') as file: df_fastANI_byGenomePair_SA = pickle.load(file); print(\"  df_fastANI_byGenomePair_SA\")\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/RESULTS/df_fastANI_byGenomePair_EC.pkl', 'rb') as file: df_fastANI_byGenomePair_EC = pickle.load(file); print(\"  df_fastANI_byGenomePair_EC\")\n",
    "print(\"\\n>>> Complete Dataframes available (fastANI results Per Genome Pair):\")\n",
    "print('df_fastANI_byGenomePair_combinedALL')\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/df_fastANI_byGenomePair_combinedALL.pkl', 'rb') as file: df_fastANI_byGenomePair_combinedALL = pickle.load(file); print(\"  df_fastANI_byGenomePair_combinedALL\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d378f137-3a58-472b-bbfc-15189cbde7b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 05 -- Make master geneome-pairs dataframe for filtering down fastANI output\n",
    "\n",
    "fastANI was run on some clades pre-filtering.  This means we need to filter down the list before we process it and update the species dataframe.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d733894-b5fa-4c67-bb19-93cf8ea469a1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Part 1: Filter down original set of fastANI results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8af1b1aa-62e2-4e24-b5b4-35d8568c9852",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "step01\n",
      "step02\n",
      "step03\n",
      "CPU times: user 4min 18s, sys: 37.6 s, total: 4min 55s\n",
      "Wall time: 4min 56s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------    \n",
    "# Starting with the data from Stage2, generate all valid genome pairs\n",
    "#     Note: Shouldn't crash notebook now that Stage2 exists, can also do in batches (see commented out code below)\n",
    "#-------------------------\n",
    "#     Input: df_gtdb_species_data_filteredStage2\n",
    "#     Output: valid_geneone_pairs_all\n",
    "#-------------------------\n",
    "# Last runtime: 4min 56s\n",
    "from itertools import combinations\n",
    "\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_species_data_filteredStage2.pkl', 'rb') as file: df_gtdb_species_data_filteredStage2 = pickle.load(file)\n",
    "\n",
    "#-- Previous verison: Had to do this in batches with the top 4 largest seperated as they crashed the notebook\n",
    "#df = df_gtdb_species_data_filteredStage2[4:].copy()  #Set05 (everything else)\n",
    "#df = df_gtdb_species_data_filteredStage2[3:4].copy() #Set04 (s__Salmonella_enterica--RS_GCF_000006945.2 - 11456 genomes)\n",
    "#df = df_gtdb_species_data_filteredStage2[2:3].copy() #Set03 (s__Staphylococcus_aureus - 14641 genomes)\n",
    "#df = df_gtdb_species_data_filteredStage2[1:2].copy() #Set02 (s__Klebsiella_pneumoniae - 14338 genomes)\n",
    "#df = df_gtdb_species_data_filteredStage2[0:1].copy() #Set01 (EC)\n",
    "#-- New version: can handle all data\n",
    "df = df_gtdb_species_data_filteredStage2\n",
    "\n",
    "# Function to generate pairwise combinations\n",
    "def pairwise_combinations(item_list):\n",
    "    return [tuple(sorted(pair)) for pair in combinations(item_list, 2)]\n",
    "\n",
    "# Apply the function to the 'Items' column\n",
    "df['Pairwise'] = df['Clustered genomes retained Stage2'].apply(pairwise_combinations)\n",
    "\n",
    "# Concatenate all pairwise combinations into a single list\n",
    "result = [pair for pairs in df['Pairwise'] for pair in pairs]\n",
    "del df\n",
    "print(f\"Total number of valid genome pairs: {len(result)}\")\n",
    "\n",
    "# Save to a file\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/valid_geneone_pairs_all_atStage2.pkl', 'wb') as file:\n",
    "    pickle.dump(result, file)\n",
    "del result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dcbd3467-8bcb-4ddb-ad3f-8d53ccadae2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Pairwise Combinations according to species cluster file: 537,061,532\n",
      "Total Pairwise Combinations according to valid pairs list:     537,061,532\n",
      "\n",
      "Value match, the world is good\n",
      "\n",
      "CPU times: user 38.8 s, sys: 19.4 s, total: 58.1 s\n",
      "Wall time: 1min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#-------------------------    \n",
    "# Validate the size of the valid genome-pair list is right by doing the math on the species clusters dataframe\n",
    "#     Input: df_gtdb_species_data_filteredStage2, valid_geneone_pairs_all\n",
    "#     Output: n/a (just prints)\n",
    "#-------------------------\n",
    "# Last runtime: 1min \n",
    "\n",
    "#--  \n",
    "# Compute how many pairs are to be expected (post-filtering Stage2) for a specific clade only\n",
    "#   Note: Selfs are not included as final genepair dataframe only uses the txt_ column\n",
    "#--\n",
    "from math import comb\n",
    "\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_species_data_filteredStage2.pkl', 'rb') as file: df_gtdb_species_data_filteredStage2 = pickle.load(file)\n",
    "df = df_gtdb_species_data_filteredStage2\n",
    "\n",
    "# Function to calculate total pairwise combinations for a list\n",
    "def total_pairwise_combinations(item_list):\n",
    "    n = item_list\n",
    "    return comb(n+1, 2) - n #all pairs w/o r.t. order and no selfs\n",
    "\n",
    "# Apply the function to the 'Items' column and create a new 'Pairwise_Total' column\n",
    "df['Pairwise_Total'] = df['No. clustered genomes Filtered Stage2'].apply(total_pairwise_combinations)\n",
    "\n",
    "# Sum the total pairwise combinations\n",
    "total_sum = df['Pairwise_Total'].sum()\n",
    "\n",
    "print(f\"Total Pairwise Combinations according to species cluster file: {total_sum:,}\")\n",
    "\n",
    "\n",
    "# Get length of valid pairs calculated above\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/valid_geneone_pairs_all_atStage2.pkl', 'rb') as file: valid_geneone_pairs_all = pickle.load(file)\n",
    "number_of_marked_valid_pairs = len(valid_geneone_pairs_all)\n",
    "del valid_geneone_pairs_all\n",
    "print(f\"Total Pairwise Combinations according to valid pairs list:     {number_of_marked_valid_pairs:,}\")\n",
    "\n",
    "\n",
    "#--  \n",
    "# Compare\n",
    "#--\n",
    "if total_sum == number_of_marked_valid_pairs: print(\"\\nValue match, the world is good\\n\")\n",
    "else: print(f\"\\nThe counts are off by {total_sum-number_of_marked_valid_pairs}, the world is not good\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6336f4c0-5342-4dd2-9eae-c1fe76054a09",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Pairwise Combinations: 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2356682/1525016944.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['Pairwise_Total'] = df['No. clustered genomes Filtered Stage2'].apply(total_pairwise_combinations)\n"
     ]
    }
   ],
   "source": [
    "#-------------------------    \n",
    "# Optional! For inspection only\n",
    "#\n",
    "# Compute how many pairs are to be expected (post-filtering Stage2) for a specific clade only (if needed)\n",
    "#   Note: Selfs are not included as final genepair dataframe only uses the txt_ column\n",
    "#-------------------------\n",
    "from math import comb\n",
    "\n",
    "CLADE_OF_INTEREST='s__UBA184_sp021817245--GB_GCA_021817245.1'\n",
    "#--------\n",
    "\n",
    "df = df_gtdb_species_data_filteredStage2[df_gtdb_species_data_filteredStage2['clade_name']==CLADE_OF_INTEREST]\n",
    "\n",
    "# Function to calculate total pairwise combinations for a list\n",
    "def total_pairwise_combinations(item_list):\n",
    "    n = item_list\n",
    "    return comb(n+1, 2) - n #all pairs w/o r.t. order and no selfs\n",
    "\n",
    "# Apply the function to the 'Items' column and create a new 'Pairwise_Total' column\n",
    "df['Pairwise_Total'] = df['No. clustered genomes Filtered Stage2'].apply(total_pairwise_combinations)\n",
    "\n",
    "# Sum the total pairwise combinations\n",
    "total_sum = df['Pairwise_Total'].sum()\n",
    "\n",
    "print(\"Total Pairwise Combinations:\", total_sum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3891248-b9bc-4e54-baf2-4faf4cc1d6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# NEW WIP\n",
    "#-------------------------    \n",
    "# Filter down fastANI results and validate counts\n",
    "#     Input: df_genepairs_ANIvAF, valid_geneone_pairs_all_atStage2\n",
    "#     Output: df_geneomepairs_ANIvAF\n",
    "#-------------------------\n",
    "# Last runtime: \n",
    "\n",
    "# Read in the original results (no filtering applied) and make a key column to merge with the valid pairs list\n",
    "# TODO---- update file name and path (df_genepairs_ANIvAF)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/STEP02_formatting_fastANI/df_fastANI_byGenomePair_combinedALL.pkl', 'rb') as file: df_genepairs_ANIvAF = pickle.load(file)\n",
    "df_genepairs_ANIvAF['key']= df_genepairs_ANIvAF[['Genome1', 'Genome2']].apply(lambda x: '_'.join(sorted([col for col in x])), axis=1)\n",
    "\n",
    "# Merge the valid pairs with the original data to perform filtering\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/valid_geneone_pairs_all_atStage2.pkl', 'rb') as file: valid_geneone_pairs_all = pickle.load(file)\n",
    "## Or Read in the valid genome pairs for the orignal set only\n",
    "#with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/temp/df_valid_genepairs_originalSet_non58.pkl', 'rb') as file: df_valid_genepairs_originalSet_non58 = pickle.load(file)\n",
    "\n",
    "# Merge\n",
    "df_merged = valid_geneone_pairs_all.merge(df_genepairs_ANIvAF, on='key', how='left')\n",
    "\n",
    "# Check for missing values\n",
    "# TODO---- Is this the only check I need to make sure I have all the information I need?  I think so...\n",
    "missing_matches = df_merged[df_merged['column_from_df_genepairs_ANIvAF'].isna()]\n",
    "if missing_matches.empty:\n",
    "    print(\"All rows from valid_geneone_pairs_all found a match in df_genepairs_ANIvAF\")\n",
    "else:\n",
    "    print(\"There are missing matches. Check the missing_matches dataframe for details.\")\n",
    "    print(missing_matches)\n",
    "\n",
    "# Save to a file\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/df_fastANI_byGenomePair_combinedALL_ValidOnly_atStage2_ANIvAF.pkl', 'wb') as file: pickle.dump(df_merged, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2987ec-0f28-4f30-91b2-7d873cf42653",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Part2 : Create Stage3 filtered species_data dataframe\n",
    "\n",
    "Was running this on NERSC with both sets of genenome pair information (see `/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/filter_down_species_cluster_all_clades.py`).\n",
    "\n",
    "Thus the below is outdated, but can be used to test new filter limits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0c68091a-a4fb-4f1f-90eb-abc4820f7b09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows to be processed: 14953\n",
      "Sample of dataframe:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome1</th>\n",
       "      <th>Genome2</th>\n",
       "      <th>ANI</th>\n",
       "      <th>AF</th>\n",
       "      <th>clade_name</th>\n",
       "      <th>Genome1_size</th>\n",
       "      <th>Genome2_size</th>\n",
       "      <th>min_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_902606795.1</td>\n",
       "      <td>GCA_003209225.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.989899</td>\n",
       "      <td>s__AAA240-E13_sp003209225--GB_GCA_003209225.1</td>\n",
       "      <td>966640</td>\n",
       "      <td>966640</td>\n",
       "      <td>966640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_902608355.1</td>\n",
       "      <td>GCA_003211695.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>s__AAA240-E13_sp003211695--GB_GCA_003211695.1</td>\n",
       "      <td>1138562</td>\n",
       "      <td>1138562</td>\n",
       "      <td>1138562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_902528115.1</td>\n",
       "      <td>GCA_003282145.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.996947</td>\n",
       "      <td>s__AAA536-G10_sp003282145--GB_GCA_003282145.1</td>\n",
       "      <td>2044868</td>\n",
       "      <td>2044868</td>\n",
       "      <td>2044868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_902760045.1</td>\n",
       "      <td>GCA_900320825.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.981994</td>\n",
       "      <td>s__AC2028_sp900320825--GB_GCA_900320825.1</td>\n",
       "      <td>3029516</td>\n",
       "      <td>3029516</td>\n",
       "      <td>3029516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Genome1          Genome2    ANI        AF  \\\n",
       "0  GCA_902606795.1  GCA_003209225.1  100.0  0.989899   \n",
       "1  GCA_902608355.1  GCA_003211695.1  100.0  1.000000   \n",
       "2  GCA_902528115.1  GCA_003282145.1  100.0  0.996947   \n",
       "3  GCA_902760045.1  GCA_900320825.1  100.0  0.981994   \n",
       "\n",
       "                                      clade_name  Genome1_size  Genome2_size  \\\n",
       "0  s__AAA240-E13_sp003209225--GB_GCA_003209225.1        966640        966640   \n",
       "1  s__AAA240-E13_sp003211695--GB_GCA_003211695.1       1138562       1138562   \n",
       "2  s__AAA536-G10_sp003282145--GB_GCA_003282145.1       2044868       2044868   \n",
       "3      s__AC2028_sp900320825--GB_GCA_900320825.1       3029516       3029516   \n",
       "\n",
       "   min_size  \n",
       "0    966640  \n",
       "1   1138562  \n",
       "2   2044868  \n",
       "3   3029516  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of removed genomes: 6794\n",
      "Total number of times a genome was already removed: 8180\n",
      "Total number of clades dropped 779\n"
     ]
    }
   ],
   "source": [
    "# This currently only runs on the originalSet, see NERSC for running on both sets (all data)\n",
    "\n",
    "#-------------------------    \n",
    "# Setup dataframe to iterate on\n",
    "#     Input: df_original_genepairs_ANIvAF_Stage2Filtered\n",
    "#     Output: df_gtdb_species_data_filteredStage3 (more columns)\n",
    "#     Dependencies: df_gtdb_metadata_filteredStage2 (for clade name), df_master_filter_keys\n",
    "#-------------------------\n",
    "# Last runtime: quick\n",
    "\n",
    "# Load in a genome pair dataframe to work off of\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/df_original_genepairs_ANIvAF_Stage2Filtered_saved_redosAdded.pkl', 'rb') as file: df_original_genepairs_ANIvAF_Stage2Filtered_saved_redosAdded = pickle.load(file)\n",
    "df = df_original_genepairs_ANIvAF_Stage2Filtered_saved_redosAdded.copy()\n",
    "del df_original_genepairs_ANIvAF_Stage2Filtered_saved_redosAdded\n",
    "\n",
    "# Set the filter\n",
    "FILTER=99.9999\n",
    "df = df[df['ANI']>=FILTER]\n",
    "\n",
    "# Add in clade_names and genome size to reference df\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_metadata_filteredStage2.pkl', 'rb') as file: df_gtdb_metadata_filteredStage2 = pickle.load(file)\n",
    "df_gtdb_metadata_filteredStage2['clade_name']=df_gtdb_metadata_filteredStage2['gtdb_taxonomy_species']+\"--\"+df_gtdb_metadata_filteredStage2['gtdb_genome_representative']\n",
    "df_gtdb_metadata_filteredStage2['clade_name'].replace(' ', '_', regex=True, inplace=True)\n",
    "df_gtdb_metadata_filteredStage2=df_gtdb_metadata_filteredStage2[['accession','clade_name','genome_size']]\n",
    "df_gtdb_metadata_filteredStage2['accession']=df_gtdb_metadata_filteredStage2['accession'].str[3:]\n",
    "\n",
    "# Add in clade_names and genome size to df\n",
    "# Merge Genome1 information & clade_name\n",
    "df = pd.merge(df, df_gtdb_metadata_filteredStage2[['accession', 'clade_name', 'genome_size']], how='left', left_on='Genome1', right_on='accession')\n",
    "df = df.rename(columns={'genome_size': 'Genome1_size'})\n",
    "# Merge Genome2 information\n",
    "df = pd.merge(df, df_gtdb_metadata_filteredStage2[['accession', 'genome_size']], how='left', left_on='Genome2', right_on='accession')\n",
    "df = df.rename(columns={'genome_size': 'Genome2_size'})\n",
    "# Clean up\n",
    "df = df.drop(['accession_x', 'accession_y'], axis=1, errors='ignore')\n",
    "\n",
    "# Make a new column for the min size (by row/axis=1)\n",
    "df['min_size']=df[['Genome1_size', 'Genome2_size']].min(axis=1)\n",
    "\n",
    "# Print stuff\n",
    "print(f\"Number of rows to be processed: {len(df)}\")\n",
    "print(\"Sample of dataframe:\")\n",
    "df[:4]\n",
    "\n",
    "\n",
    "#-------------------------    \n",
    "# Perform filtering of individual genomes\n",
    "#-------------------------\n",
    "# Last runtime: \n",
    "\n",
    "# Setup\n",
    "set_of_removed_genomes = set()\n",
    "count_number_of_already_removed=0\n",
    "df.sort_values(by='min_size', inplace=True, ignore_index=True)\n",
    "\n",
    "\n",
    "# Iterate\n",
    "for index, row in df.iterrows():\n",
    "    #if index%10000==0: print(index)\n",
    "    if row['Genome1'] in set_of_removed_genomes or row['Genome2'] in set_of_removed_genomes:\n",
    "        count_number_of_already_removed+=1\n",
    "    if row['Genome1_size']<row['Genome2_size']:\n",
    "        set_of_removed_genomes.add(row['Genome1'])\n",
    "    else:\n",
    "        set_of_removed_genomes.add(row['Genome2'])\n",
    "        \n",
    "# Finish up\n",
    "print(f\"Total number of removed genomes: {len(set_of_removed_genomes)}\")\n",
    "print(f\"Total number of times a genome was already removed: {count_number_of_already_removed}\")\n",
    "\n",
    "# Remove genomes from species data dataframe to see how many clades are lost\n",
    "\n",
    "\n",
    "#-------------------------    \n",
    "# Clades removed\n",
    "#-------------------------\n",
    "# ---- Part 1, update master keys dataframe\n",
    "# Read in any source data needed\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/df_master_filter_keys.pkl', 'rb') as file: df_master_filter_keys = pickle.load(file)\n",
    "#-- Filter based on clade size (FilterE)\n",
    "df_master_filter_keys['FilterE1'] = np.where(df_master_filter_keys['genome_ID'].str[3:].isin(list(set_of_removed_genomes)), 1, 0)\n",
    "\n",
    "# ---- Part 2, merge with species dataframe\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_species_data_filteredStage2.pkl', 'rb') as file: df_gtdb_species_data_filteredStage2 = pickle.load(file)\n",
    "df_gtdb_species_data_filteredStage3 = df_gtdb_species_data_filteredStage2.copy()\n",
    "del df_gtdb_species_data_filteredStage2\n",
    "accessions_to_filter_out_set = set(df_master_filter_keys[(df_master_filter_keys['FilterE1']==1)]['genome_ID'].to_list())\n",
    "\n",
    "# Step 1: Filter the list of genomes in each clade\n",
    "# Create a log of any genomes to be removed in new column\n",
    "df_gtdb_species_data_filteredStage3[\"Clustered genomes removed Stage3\"] = (\n",
    "    df_gtdb_species_data_filteredStage3[\"Clustered genomes retained Stage2\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "# Update the \"Clustered genomes\" column after applying the filters\n",
    "df_gtdb_species_data_filteredStage3[\"Clustered genomes retained\"] = (\n",
    "    df_gtdb_species_data_filteredStage3[\"Clustered genomes retained Stage2\"]\n",
    "    .explode()\n",
    "    .loc[lambda x: ~x.isin(accessions_to_filter_out_set)]\n",
    "    .groupby(level=0)\n",
    "    .agg(list)\n",
    ")\n",
    "\n",
    "# Step 2: Calculate the length of the updated \"Clustered genomes\" list and create a new column, rename accordingly\n",
    "df_gtdb_species_data_filteredStage3[\"No. clustered genomes Filtered Stage3\"] = df_gtdb_species_data_filteredStage3[\"Clustered genomes retained\"].apply(lambda x: len(x) if isinstance(x, list) else 0)\n",
    "df_gtdb_species_data_filteredStage3 = df_gtdb_species_data_filteredStage3.rename(columns={'No. clustered genomes': 'No. clustered genomes Filter Stage2', 'Clustered genomes retained': 'Clustered genomes retained Stage3'})\n",
    "\n",
    "# Step 3: Filter again w.r.t. size\n",
    "new_accessions_to_filter_out = df_gtdb_species_data_filteredStage3[df_gtdb_species_data_filteredStage3[\"No. clustered genomes Filtered Stage1\"] < filter_min_clade_size][\"Clustered genomes retained Stage3\"]\n",
    "filtered_list = [x for x in new_accessions_to_filter_out.to_list() if str(x) != 'nan']\n",
    "new_accessions_to_filter_out = [item for sublist in filtered_list for item in sublist]\n",
    "#df_gtdb_species_data_filteredStage1 = df_gtdb_species_data_filteredStage1[df_gtdb_species_data_filteredStage1['Filtered No. clustered genomes'] >= filter_min_clade_size]\n",
    "\n",
    "# Length Check\n",
    "clades_dropped=len(df_gtdb_species_data_filteredStage3[df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage3']<filter_min_clade_size])\n",
    "print(f\"Total number of clades dropped {clades_dropped}\")\n",
    "    \n",
    "# Step 4: Finish filtering the dataframe (drop any < min size) and save\n",
    "df_gtdb_species_data_filteredStage3 = df_gtdb_species_data_filteredStage3[df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage3'] >= filter_min_clade_size]\n",
    "\n",
    "# Save files\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/df_gtdb_species_data_filteredStage3.pkl', 'wb') as file: pickle.dump(df_gtdb_species_data_filteredStage3, file)\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/df_master_filter_keys_filteredStage3.pkl', 'wb') as file: pickle.dump(df_master_filter_keys, file)\n",
    "\n",
    "# Print out many rows changed as a sanity check\n",
    "row_change_count = len(df_gtdb_species_data_filteredStage3[~(df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage2']==df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage3'])])\n",
    "print(f\"Total number of clades changed {row_change_count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36638d59-6187-4657-91ec-95cc0b058e01",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 06 -- Setup & Run mOTUpan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71327d43-ed08-4b1d-a65d-cd5ec4736343",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Part 1 - Generate clade lists for mOTUpan runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d31c9361-fc6f-45cd-9035-56c628e65b5a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#-------------------------    \n",
    "# Generate clade list files to run mOTUpan\n",
    "#     Input: df_original_genepairs_ANIvAF_Stage2Filtered\n",
    "#     Output: df_gtdb_species_data_filteredStage3 (more columns)\n",
    "#     Dependencies: df_gtdb_metadata_filteredStage2 (for clade name), df_master_filter_keys\n",
    "#-------------------------\n",
    "# Last runtime: quick\n",
    "\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/df_gtdb_species_data_filteredStage3_99pt9ANIthresh.pkl', 'rb') as file: df_gtdb_species_data_filteredStage3 = pickle.load(file)\n",
    "\n",
    "# Split into two files for batching (clade sizes less than 500 and clade sizes greater than or equal to 500)\n",
    "dir_out = \"/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/\"\n",
    "file_out_LT = \"clade_list_xANIthresh_sizeAllLT500.txt\"\n",
    "file_out_GT = \"clade_list_xANIthresh_sizeGTE500.txt\"\n",
    "\n",
    "# Sort by size so the smaller runs finish first\n",
    "df_gtdb_species_data_filteredStage3=df_gtdb_species_data_filteredStage3.sort_values('No. clustered genomes Filtered Stage3')\n",
    "\n",
    "# Get list of clade names and distribute\n",
    "clade_list_LT = df_gtdb_species_data_filteredStage3[df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage3'] < 500]['clade_name'].tolist()\n",
    "clade_list_GT = df_gtdb_species_data_filteredStage3[df_gtdb_species_data_filteredStage3['No. clustered genomes Filtered Stage3'] >= 500]['clade_name'].tolist()\n",
    "print(f\"Clades less than 500 in size: {len(clade_list_LT)}\")\n",
    "clade_list_LT[:2]\n",
    "print(f\"\\nClades greater than or equal to 500 in size: {len(clade_list_GT)}\")\n",
    "clade_list_GT[:2]\n",
    "\n",
    "# Save both lists\n",
    "with open(dir_out+file_out_LT, 'w') as file: \n",
    "    for item in clade_list_LT: file.write(f\"{item}\\n\")\n",
    "with open(dir_out+file_out_GT, 'w') as file: \n",
    "    for item in clade_list_GT: file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ced8f358-fad5-4563-999f-e6bd8764dc4e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Part2 - Generate faa pathlists for each clade\n",
    "\n",
    "Also running on NERSC (see generate_genome_pathlists_faa.py in /filtering_stage3)\n",
    "\n",
    "Note taking too long on interactive nodes (not done in 4 hours, but can get to run in 15 min on login node)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "21bdf218-51d7-4ea5-bf91-f555ab9991ca",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#-------------------------    \n",
    "# Generate pathlists to faa files\n",
    "#     Input: df_gtdb_species_data_filteredStage3\n",
    "#     Output: clade_path_lists_filtered_stage3/genome_pathlist_CLADE.txt\n",
    "# Note: copied from script on NERSC\n",
    "#-------------------------\n",
    "\n",
    "print(\"WARNING: This cell will exit immediately to prevent accidental execution - unsure if jup can handle it, please run on login node for ~15 min execution.\")\n",
    "raise KeyboardInterrupt(\"Cell execution stopped by user.\")\n",
    "\n",
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/df_gtdb_species_data_filteredStage3.pkl', 'rb') as file: df_gtdb_species_data_filtered = pickle.load(file)\n",
    "\n",
    "\n",
    "# Generate pathlists (paths to fnas for a clade) for a LIST of clades\n",
    "#-------------------------\n",
    "\n",
    "# Get list of clades to generate pathlists for using desired criteria \n",
    "#to_gen=df_gtdb_species_data_filtered[(df_gtdb_species_data_filtered['Filtered No. clustered genomes']>500) & (df_gtdb_species_data_filtered['Filtered No. clustered genomes']<32000)]['clade_name'].to_list()\n",
    "to_gen=df_gtdb_species_data_filtered[:]['clade_name'].to_list()\n",
    "\n",
    "# Iterate through each clade\n",
    "for clade in to_gen:\n",
    "    #print(clade)\n",
    "    \n",
    "    # Custom do a single one (e.g. need to drop a genome)\n",
    "    clade_name=clade #\"s__Escherichia_coli--RS_GCF_003697165.2\"\n",
    "    outpath=\"/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/clade_path_lists_filtered_stage3\"\n",
    "    base_directory = \"/global/cfs/cdirs/kbase/jungbluth/Projects/Project_Pangenome_GTDB/GTDB_r214_by_spcluster/\"\n",
    "    #base_directory_out = \"GTDB_v214_download/ftp.ncbi.nlm.nih.gov/genomes/all/\"\n",
    "    row=df_gtdb_species_data_filtered[df_gtdb_species_data_filtered['clade_name'] == clade_name]\n",
    "\n",
    "    outfile = outpath + \"/genome_pathlist_\" + clade_name + \".txt\"\n",
    "    genomes = row['Clustered genomes retained Stage3'].iloc[0]\n",
    "    if ~(len(genomes) == row['No. clustered genomes Filtered Stage3'].iloc[0]):\n",
    "        print(\"Something went wrong\")\n",
    "        exit(1)\n",
    "    paths = []\n",
    "\n",
    "    for genome in genomes:\n",
    "        genome = genome[3:]\n",
    "        #print(genome)\n",
    "\n",
    "        # Construct the subdirectory path based on the given string\n",
    "        subdirectory_path = os.path.join(base_directory, clade)\n",
    "\n",
    "        # Initialize a variable to count matching files\n",
    "        matching_file_count = 0\n",
    "        matching_file_path = None\n",
    "\n",
    "        # Iterate through subdirectories and look for .fna.gz files\n",
    "        for file in os.listdir(subdirectory_path):\n",
    "            if file.endswith(\".faa.gz\") and genome in file:\n",
    "                matching_file_count += 1\n",
    "                matching_file_path = os.path.join(base_directory, file)\n",
    "                #print(matching_file_path)\n",
    "\n",
    "                # If more than one matching file is found, raise an error\n",
    "                if matching_file_count > 1:\n",
    "                    print(\"\\n\")\n",
    "                    print(row['clade_name'])\n",
    "                    print(genome)\n",
    "                    print(matching_file_path)\n",
    "                    raise ValueError(\"Multiple matching files found\")\n",
    "\n",
    "        # Check the result and print the matching file path if one is found\n",
    "        if matching_file_count == 0:\n",
    "            print(row['clade_name'])\n",
    "            print(genome)\n",
    "            raise ValueError(\"No matching files found\")\n",
    "        elif matching_file_count == 1:\n",
    "            paths.append(matching_file_path)\n",
    "            #print(\"Found:\", matching_file_path)\n",
    "\n",
    "    type(paths)\n",
    "    # Join the strings with newline characters and write to the file\n",
    "    with open(outfile, 'w') as file:\n",
    "        file.write('\\n'.join(paths))\n",
    "        file.write('\\n')\n",
    "        paths = []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83820d39-b3bc-4cb1-9c13-a7710c520453",
   "metadata": {},
   "source": [
    "## Run mOTUpan\n",
    "\n",
    "On NERSC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b311782-8fd2-45f5-8e14-500183477aa3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# -- STEP 07 -- Processing mOTUpan results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f2b7e25-eb7a-440c-8bcc-ef01bd2b6a2b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "307919"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage3/df_gtdb_species_data_filteredStage3.pkl', 'rb') as file: df_gtdb_species_data_filteredStage3 = pickle.load(file)\n",
    "#df_gtdb_species_data_filteredStage3\n",
    "list_of_all_valid_genomes_Stage3 = df_gtdb_species_data_filteredStage3['Clustered genomes retained Stage3'].explode().tolist()\n",
    "len(list_of_all_valid_genomes_Stage3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6faa34ae-16e0-42b9-9d56-dbdefb5b2823",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GB_GCA_000193895.2', 'GB_GCA_000194495.2']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_all_valid_genomes_Stage3[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "bb984e9d-98a3-4fc3-a2a1-dbe05b141d88",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'pangenome_utils' from '/global/cfs/cdirs/kbase/ke_prototype/mcashman/scratch_notebooks/pangenome_utils.py'>"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(pg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ddfff257-7bcf-45a7-afe7-9bcdb56a1464",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade_name</th>\n",
       "      <th>checkm_average</th>\n",
       "      <th>checkm_median</th>\n",
       "      <th>checkm_min</th>\n",
       "      <th>checkm_max</th>\n",
       "      <th>contig_average</th>\n",
       "      <th>contig_median</th>\n",
       "      <th>contig_min</th>\n",
       "      <th>contig_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>s__0-14-0-80-60-11_sp018897875--GB_GCA_0188978...</td>\n",
       "      <td>89.226667</td>\n",
       "      <td>89.870</td>\n",
       "      <td>87.35</td>\n",
       "      <td>90.46</td>\n",
       "      <td>540.666667</td>\n",
       "      <td>537.0</td>\n",
       "      <td>442</td>\n",
       "      <td>643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s__0-14-3-00-41-53_sp002780895--GB_GCA_0027808...</td>\n",
       "      <td>93.844286</td>\n",
       "      <td>96.590</td>\n",
       "      <td>85.09</td>\n",
       "      <td>96.59</td>\n",
       "      <td>200.285714</td>\n",
       "      <td>230.0</td>\n",
       "      <td>71</td>\n",
       "      <td>322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s__01-FULL-36-15b_sp001782035--GB_GCA_001782035.1</td>\n",
       "      <td>60.890000</td>\n",
       "      <td>60.890</td>\n",
       "      <td>60.89</td>\n",
       "      <td>60.89</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s__01-FULL-44-24b_sp001793235--GB_GCA_001793235.1</td>\n",
       "      <td>56.105000</td>\n",
       "      <td>56.105</td>\n",
       "      <td>55.61</td>\n",
       "      <td>56.60</td>\n",
       "      <td>30.500000</td>\n",
       "      <td>30.5</td>\n",
       "      <td>29</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s__01-FULL-45-10b_sp001804205--GB_GCA_001804205.1</td>\n",
       "      <td>88.375000</td>\n",
       "      <td>88.375</td>\n",
       "      <td>85.42</td>\n",
       "      <td>91.33</td>\n",
       "      <td>246.500000</td>\n",
       "      <td>246.5</td>\n",
       "      <td>202</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27698</th>\n",
       "      <td>s__Zunongwangia_profunda--RS_GCF_000023465.1</td>\n",
       "      <td>87.163077</td>\n",
       "      <td>88.100</td>\n",
       "      <td>57.10</td>\n",
       "      <td>99.62</td>\n",
       "      <td>387.153846</td>\n",
       "      <td>423.0</td>\n",
       "      <td>1</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27699</th>\n",
       "      <td>s__Zwartia_sp903958565--GB_GCA_903958565.1</td>\n",
       "      <td>97.966000</td>\n",
       "      <td>98.340</td>\n",
       "      <td>94.67</td>\n",
       "      <td>99.64</td>\n",
       "      <td>454.000000</td>\n",
       "      <td>417.0</td>\n",
       "      <td>407</td>\n",
       "      <td>587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27700</th>\n",
       "      <td>s__Zymobacter_palmae--RS_GCF_003610015.1</td>\n",
       "      <td>95.200000</td>\n",
       "      <td>95.200</td>\n",
       "      <td>95.19</td>\n",
       "      <td>95.21</td>\n",
       "      <td>36.500000</td>\n",
       "      <td>36.5</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27701</th>\n",
       "      <td>s__Zymomonas_mobilis--RS_GCF_000175255.2</td>\n",
       "      <td>99.747308</td>\n",
       "      <td>100.000</td>\n",
       "      <td>94.11</td>\n",
       "      <td>100.00</td>\n",
       "      <td>18.846154</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27702</th>\n",
       "      <td>s__Zymomonas_pomaceae--RS_GCF_000218875.1</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>17.500000</td>\n",
       "      <td>17.5</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27703 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              clade_name  checkm_average  \\\n",
       "0      s__0-14-0-80-60-11_sp018897875--GB_GCA_0188978...       89.226667   \n",
       "1      s__0-14-3-00-41-53_sp002780895--GB_GCA_0027808...       93.844286   \n",
       "2      s__01-FULL-36-15b_sp001782035--GB_GCA_001782035.1       60.890000   \n",
       "3      s__01-FULL-44-24b_sp001793235--GB_GCA_001793235.1       56.105000   \n",
       "4      s__01-FULL-45-10b_sp001804205--GB_GCA_001804205.1       88.375000   \n",
       "...                                                  ...             ...   \n",
       "27698       s__Zunongwangia_profunda--RS_GCF_000023465.1       87.163077   \n",
       "27699         s__Zwartia_sp903958565--GB_GCA_903958565.1       97.966000   \n",
       "27700           s__Zymobacter_palmae--RS_GCF_003610015.1       95.200000   \n",
       "27701           s__Zymomonas_mobilis--RS_GCF_000175255.2       99.747308   \n",
       "27702          s__Zymomonas_pomaceae--RS_GCF_000218875.1      100.000000   \n",
       "\n",
       "       checkm_median  checkm_min  checkm_max  contig_average  contig_median  \\\n",
       "0             89.870       87.35       90.46      540.666667          537.0   \n",
       "1             96.590       85.09       96.59      200.285714          230.0   \n",
       "2             60.890       60.89       60.89        7.000000            7.0   \n",
       "3             56.105       55.61       56.60       30.500000           30.5   \n",
       "4             88.375       85.42       91.33      246.500000          246.5   \n",
       "...              ...         ...         ...             ...            ...   \n",
       "27698         88.100       57.10       99.62      387.153846          423.0   \n",
       "27699         98.340       94.67       99.64      454.000000          417.0   \n",
       "27700         95.200       95.19       95.21       36.500000           36.5   \n",
       "27701        100.000       94.11      100.00       18.846154            5.0   \n",
       "27702        100.000      100.00      100.00       17.500000           17.5   \n",
       "\n",
       "       contig_min  contig_max  \n",
       "0             442         643  \n",
       "1              71         322  \n",
       "2               5           9  \n",
       "3              29          32  \n",
       "4             202         291  \n",
       "...           ...         ...  \n",
       "27698           1         742  \n",
       "27699         407         587  \n",
       "27700           2          71  \n",
       "27701           1         249  \n",
       "27702           3          32  \n",
       "\n",
       "[27703 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtered_stage2/df_gtdb_metadata_filteredStage2.pkl', 'rb') as file: df_gtdb_metadata_filteredStage2 = pickle.load(file)\n",
    "df_gtdb_metadata_filteredStage2['clade_name']=df_gtdb_metadata_filteredStage2['gtdb_taxonomy_species']+\"--\"+df_gtdb_metadata_filteredStage2['gtdb_genome_representative']\n",
    "df_gtdb_metadata_filteredStage2['clade_name'].replace(' ', '_', regex=True, inplace=True)\n",
    "\n",
    "col_of_interest = ['accession','clade_name','checkm_completeness','checkm_marker_count','contig_count','gtdb_taxonomy_species']\n",
    "df = df_gtdb_metadata_filteredStage2[df_gtdb_metadata_filteredStage2['accession'].isin(list_of_all_valid_genomes_Stage3)][col_of_interest]\n",
    "\n",
    "grouped_df = df.groupby('clade_name').agg(\n",
    "    checkm_average=pd.NamedAgg(column='checkm_completeness', aggfunc='mean'),\n",
    "    checkm_median=pd.NamedAgg(column='checkm_completeness', aggfunc='median'),\n",
    "    checkm_min=pd.NamedAgg(column='checkm_completeness', aggfunc='min'),\n",
    "    checkm_max=pd.NamedAgg(column='checkm_completeness', aggfunc='max'),\n",
    "    contig_average=pd.NamedAgg(column='contig_count', aggfunc='mean'),\n",
    "    contig_median=pd.NamedAgg(column='contig_count', aggfunc='median'),\n",
    "    contig_min=pd.NamedAgg(column='contig_count', aggfunc='min'),\n",
    "    contig_max=pd.NamedAgg(column='contig_count', aggfunc='max')\n",
    "    # Add more statistics columns as needed\n",
    ").reset_index()\n",
    "\n",
    "grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5f36f518-5c95-4968-aa83-c4cbc16ba4ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade_name</th>\n",
       "      <th>checkm_average</th>\n",
       "      <th>checkm_median</th>\n",
       "      <th>checkm_min</th>\n",
       "      <th>checkm_max</th>\n",
       "      <th>contig_average</th>\n",
       "      <th>contig_median</th>\n",
       "      <th>contig_min</th>\n",
       "      <th>contig_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>s__Acinetobacter_baumannii--RS_GCF_009759685.1</td>\n",
       "      <td>99.821837</td>\n",
       "      <td>99.98</td>\n",
       "      <td>57.57</td>\n",
       "      <td>100.00</td>\n",
       "      <td>138.579058</td>\n",
       "      <td>97.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2461</th>\n",
       "      <td>s__Bacillus_A_bombysepticus--RS_GCF_006384875.1</td>\n",
       "      <td>98.933842</td>\n",
       "      <td>99.18</td>\n",
       "      <td>54.46</td>\n",
       "      <td>99.69</td>\n",
       "      <td>183.734727</td>\n",
       "      <td>123.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2551</th>\n",
       "      <td>s__Bacillus_subtilis--RS_GCF_000009045.1</td>\n",
       "      <td>99.592218</td>\n",
       "      <td>99.81</td>\n",
       "      <td>74.14</td>\n",
       "      <td>100.00</td>\n",
       "      <td>43.341682</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2556</th>\n",
       "      <td>s__Bacillus_velezensis--RS_GCF_001461825.1</td>\n",
       "      <td>99.618485</td>\n",
       "      <td>99.67</td>\n",
       "      <td>86.05</td>\n",
       "      <td>100.00</td>\n",
       "      <td>36.077280</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2778</th>\n",
       "      <td>s__Bifidobacterium_longum--RS_GCF_000196555.1</td>\n",
       "      <td>99.271630</td>\n",
       "      <td>100.00</td>\n",
       "      <td>66.59</td>\n",
       "      <td>100.00</td>\n",
       "      <td>65.646296</td>\n",
       "      <td>45.5</td>\n",
       "      <td>1</td>\n",
       "      <td>975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3064</th>\n",
       "      <td>s__Bordetella_pertussis--RS_GCF_000306945.1</td>\n",
       "      <td>99.534642</td>\n",
       "      <td>100.00</td>\n",
       "      <td>93.11</td>\n",
       "      <td>100.00</td>\n",
       "      <td>194.961727</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3342</th>\n",
       "      <td>s__Brucella_melitensis--RS_GCF_000007125.1</td>\n",
       "      <td>99.095995</td>\n",
       "      <td>99.45</td>\n",
       "      <td>87.85</td>\n",
       "      <td>100.00</td>\n",
       "      <td>32.812796</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1</td>\n",
       "      <td>722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3409</th>\n",
       "      <td>s__Burkholderia_cenocepacia--RS_GCF_900446215.1</td>\n",
       "      <td>99.225651</td>\n",
       "      <td>99.73</td>\n",
       "      <td>92.74</td>\n",
       "      <td>99.87</td>\n",
       "      <td>160.485207</td>\n",
       "      <td>138.0</td>\n",
       "      <td>3</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3423</th>\n",
       "      <td>s__Burkholderia_mallei--RS_GCF_000011705.1</td>\n",
       "      <td>99.859408</td>\n",
       "      <td>100.00</td>\n",
       "      <td>95.25</td>\n",
       "      <td>100.00</td>\n",
       "      <td>128.373349</td>\n",
       "      <td>104.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3426</th>\n",
       "      <td>s__Burkholderia_multivorans--RS_GCF_000959525.1</td>\n",
       "      <td>99.853292</td>\n",
       "      <td>100.00</td>\n",
       "      <td>95.95</td>\n",
       "      <td>100.00</td>\n",
       "      <td>103.606436</td>\n",
       "      <td>83.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3450</th>\n",
       "      <td>s__Burkholderia_ubonensis--RS_GCF_902833085.1</td>\n",
       "      <td>99.815180</td>\n",
       "      <td>100.00</td>\n",
       "      <td>96.13</td>\n",
       "      <td>100.00</td>\n",
       "      <td>130.868852</td>\n",
       "      <td>111.0</td>\n",
       "      <td>3</td>\n",
       "      <td>1143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5020</th>\n",
       "      <td>s__Campylobacter_D_coli--RS_GCF_000254135.1</td>\n",
       "      <td>99.334832</td>\n",
       "      <td>99.44</td>\n",
       "      <td>94.89</td>\n",
       "      <td>99.72</td>\n",
       "      <td>58.293805</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1</td>\n",
       "      <td>577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5026</th>\n",
       "      <td>s__Campylobacter_D_jejuni--RS_GCF_001457695.1</td>\n",
       "      <td>99.234280</td>\n",
       "      <td>99.53</td>\n",
       "      <td>85.63</td>\n",
       "      <td>99.88</td>\n",
       "      <td>56.876783</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1</td>\n",
       "      <td>514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5506</th>\n",
       "      <td>s__Citrobacter_freundii--RS_GCF_011064845.1</td>\n",
       "      <td>99.772334</td>\n",
       "      <td>99.90</td>\n",
       "      <td>83.75</td>\n",
       "      <td>100.00</td>\n",
       "      <td>127.188153</td>\n",
       "      <td>89.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5535</th>\n",
       "      <td>s__Clostridioides_difficile--RS_GCF_001077535.1</td>\n",
       "      <td>98.996164</td>\n",
       "      <td>99.45</td>\n",
       "      <td>77.59</td>\n",
       "      <td>100.00</td>\n",
       "      <td>290.946621</td>\n",
       "      <td>103.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6223</th>\n",
       "      <td>s__Cronobacter_sakazakii--RS_GCF_000982825.1</td>\n",
       "      <td>99.680824</td>\n",
       "      <td>99.93</td>\n",
       "      <td>90.86</td>\n",
       "      <td>99.99</td>\n",
       "      <td>129.543430</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7409</th>\n",
       "      <td>s__Enterobacter_hormaechei_A--RS_GCF_001729745.1</td>\n",
       "      <td>99.533322</td>\n",
       "      <td>99.67</td>\n",
       "      <td>57.05</td>\n",
       "      <td>100.00</td>\n",
       "      <td>107.831635</td>\n",
       "      <td>75.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7456</th>\n",
       "      <td>s__Enterococcus_B_faecium--RS_GCF_001544255.1</td>\n",
       "      <td>99.570158</td>\n",
       "      <td>99.63</td>\n",
       "      <td>88.42</td>\n",
       "      <td>100.00</td>\n",
       "      <td>168.013818</td>\n",
       "      <td>176.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491</th>\n",
       "      <td>s__Enterococcus_faecalis--RS_GCF_000392875.1</td>\n",
       "      <td>99.528996</td>\n",
       "      <td>99.63</td>\n",
       "      <td>75.03</td>\n",
       "      <td>100.00</td>\n",
       "      <td>65.322667</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7647</th>\n",
       "      <td>s__Escherichia_coli--RS_GCF_003697165.2</td>\n",
       "      <td>99.716624</td>\n",
       "      <td>99.93</td>\n",
       "      <td>55.50</td>\n",
       "      <td>100.00</td>\n",
       "      <td>197.466824</td>\n",
       "      <td>151.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8556</th>\n",
       "      <td>s__Francisella_tularensis--RS_GCF_000008985.1</td>\n",
       "      <td>99.700458</td>\n",
       "      <td>99.82</td>\n",
       "      <td>93.88</td>\n",
       "      <td>99.93</td>\n",
       "      <td>96.071596</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1</td>\n",
       "      <td>958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9746</th>\n",
       "      <td>s__Haemophilus_influenzae--RS_GCF_001457655.1</td>\n",
       "      <td>99.650459</td>\n",
       "      <td>99.67</td>\n",
       "      <td>98.35</td>\n",
       "      <td>100.00</td>\n",
       "      <td>55.768959</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10062</th>\n",
       "      <td>s__Helicobacter_pylori--RS_GCF_900478295.1</td>\n",
       "      <td>99.388257</td>\n",
       "      <td>99.42</td>\n",
       "      <td>93.07</td>\n",
       "      <td>99.77</td>\n",
       "      <td>51.165394</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1</td>\n",
       "      <td>432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10065</th>\n",
       "      <td>s__Helicobacter_pylori_C--RS_GCF_000277405.1</td>\n",
       "      <td>99.395205</td>\n",
       "      <td>99.42</td>\n",
       "      <td>88.05</td>\n",
       "      <td>100.00</td>\n",
       "      <td>28.668309</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1</td>\n",
       "      <td>449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11881</th>\n",
       "      <td>s__Klebsiella_michiganensis--RS_GCF_002925905.1</td>\n",
       "      <td>99.754060</td>\n",
       "      <td>100.00</td>\n",
       "      <td>74.64</td>\n",
       "      <td>100.00</td>\n",
       "      <td>140.378446</td>\n",
       "      <td>108.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11885</th>\n",
       "      <td>s__Klebsiella_pneumoniae--RS_GCF_000742135.1</td>\n",
       "      <td>99.859633</td>\n",
       "      <td>100.00</td>\n",
       "      <td>70.27</td>\n",
       "      <td>100.00</td>\n",
       "      <td>126.444944</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11886</th>\n",
       "      <td>s__Klebsiella_quasipneumoniae--RS_GCF_020525925.1</td>\n",
       "      <td>99.853109</td>\n",
       "      <td>99.99</td>\n",
       "      <td>93.58</td>\n",
       "      <td>100.00</td>\n",
       "      <td>127.797753</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11890</th>\n",
       "      <td>s__Klebsiella_variicola--RS_GCF_000828055.2</td>\n",
       "      <td>99.841411</td>\n",
       "      <td>100.00</td>\n",
       "      <td>85.89</td>\n",
       "      <td>100.00</td>\n",
       "      <td>111.490798</td>\n",
       "      <td>64.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12093</th>\n",
       "      <td>s__Lacticaseibacillus_rhamnosus--RS_GCF_900636...</td>\n",
       "      <td>99.234896</td>\n",
       "      <td>99.46</td>\n",
       "      <td>91.01</td>\n",
       "      <td>100.00</td>\n",
       "      <td>89.692946</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12107</th>\n",
       "      <td>s__Lactiplantibacillus_plantarum--RS_GCF_01413...</td>\n",
       "      <td>99.145888</td>\n",
       "      <td>99.38</td>\n",
       "      <td>90.87</td>\n",
       "      <td>99.38</td>\n",
       "      <td>88.426230</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1</td>\n",
       "      <td>757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12280</th>\n",
       "      <td>s__Legionella_pneumophila--RS_GCF_000008485.1</td>\n",
       "      <td>99.908789</td>\n",
       "      <td>100.00</td>\n",
       "      <td>93.89</td>\n",
       "      <td>100.00</td>\n",
       "      <td>53.524941</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12418</th>\n",
       "      <td>s__Leptospira_interrogans--RS_GCF_900156205.1</td>\n",
       "      <td>96.455399</td>\n",
       "      <td>96.47</td>\n",
       "      <td>92.94</td>\n",
       "      <td>97.65</td>\n",
       "      <td>229.484848</td>\n",
       "      <td>216.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12763</th>\n",
       "      <td>s__Listeria_monocytogenes--RS_GCF_900187225.1</td>\n",
       "      <td>99.410407</td>\n",
       "      <td>99.45</td>\n",
       "      <td>90.66</td>\n",
       "      <td>100.00</td>\n",
       "      <td>35.770796</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1</td>\n",
       "      <td>935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12764</th>\n",
       "      <td>s__Listeria_monocytogenes_B--RS_GCF_000307025.1</td>\n",
       "      <td>99.429626</td>\n",
       "      <td>99.45</td>\n",
       "      <td>93.99</td>\n",
       "      <td>100.00</td>\n",
       "      <td>33.034321</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1</td>\n",
       "      <td>891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14627</th>\n",
       "      <td>s__Mycobacterium_abscessus--RS_GCF_000069185.1</td>\n",
       "      <td>99.899507</td>\n",
       "      <td>100.00</td>\n",
       "      <td>94.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>39.924932</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14808</th>\n",
       "      <td>s__Mycobacterium_tuberculosis--RS_GCF_000195955.2</td>\n",
       "      <td>99.784281</td>\n",
       "      <td>99.94</td>\n",
       "      <td>93.00</td>\n",
       "      <td>99.94</td>\n",
       "      <td>102.057946</td>\n",
       "      <td>77.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15027</th>\n",
       "      <td>s__Neisseria_gonorrhoeae--RS_GCF_003315235.1</td>\n",
       "      <td>99.702851</td>\n",
       "      <td>99.82</td>\n",
       "      <td>97.31</td>\n",
       "      <td>99.89</td>\n",
       "      <td>116.353249</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1</td>\n",
       "      <td>882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15030</th>\n",
       "      <td>s__Neisseria_meningitidis--RS_GCF_900638555.1</td>\n",
       "      <td>99.631975</td>\n",
       "      <td>99.75</td>\n",
       "      <td>95.48</td>\n",
       "      <td>99.83</td>\n",
       "      <td>155.097124</td>\n",
       "      <td>134.0</td>\n",
       "      <td>1</td>\n",
       "      <td>719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17272</th>\n",
       "      <td>s__Piscirickettsia_salmonis--RS_GCF_000297215.2</td>\n",
       "      <td>98.788795</td>\n",
       "      <td>98.81</td>\n",
       "      <td>98.24</td>\n",
       "      <td>98.85</td>\n",
       "      <td>34.614458</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2</td>\n",
       "      <td>547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18050</th>\n",
       "      <td>s__Proteus_mirabilis--RS_GCF_000160755.1</td>\n",
       "      <td>99.820749</td>\n",
       "      <td>100.00</td>\n",
       "      <td>75.44</td>\n",
       "      <td>100.00</td>\n",
       "      <td>161.011516</td>\n",
       "      <td>89.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18299</th>\n",
       "      <td>s__Pseudomonas_E_avellanae--RS_GCF_000444135.1</td>\n",
       "      <td>99.734828</td>\n",
       "      <td>100.00</td>\n",
       "      <td>91.01</td>\n",
       "      <td>100.00</td>\n",
       "      <td>327.241379</td>\n",
       "      <td>320.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18468</th>\n",
       "      <td>s__Pseudomonas_E_paracarnis--RS_GCF_904063055.1</td>\n",
       "      <td>99.210000</td>\n",
       "      <td>99.80</td>\n",
       "      <td>78.79</td>\n",
       "      <td>99.80</td>\n",
       "      <td>154.647727</td>\n",
       "      <td>100.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18663</th>\n",
       "      <td>s__Pseudomonas_E_viridiflava--RS_GCF_001642795.1</td>\n",
       "      <td>99.820437</td>\n",
       "      <td>100.00</td>\n",
       "      <td>96.43</td>\n",
       "      <td>100.00</td>\n",
       "      <td>90.516710</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18688</th>\n",
       "      <td>s__Pseudomonas_aeruginosa--RS_GCF_001457615.1</td>\n",
       "      <td>99.460522</td>\n",
       "      <td>99.68</td>\n",
       "      <td>81.03</td>\n",
       "      <td>100.00</td>\n",
       "      <td>163.096302</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21210</th>\n",
       "      <td>s__Salmonella_enterica--RS_GCF_000006945.2</td>\n",
       "      <td>99.922754</td>\n",
       "      <td>100.00</td>\n",
       "      <td>58.97</td>\n",
       "      <td>100.00</td>\n",
       "      <td>93.888353</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21964</th>\n",
       "      <td>s__Staphylococcus_aureus--RS_GCF_001027105.1</td>\n",
       "      <td>99.414489</td>\n",
       "      <td>99.51</td>\n",
       "      <td>83.21</td>\n",
       "      <td>99.70</td>\n",
       "      <td>61.840080</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21980</th>\n",
       "      <td>s__Staphylococcus_epidermidis--RS_GCF_006742205.1</td>\n",
       "      <td>99.604722</td>\n",
       "      <td>99.81</td>\n",
       "      <td>59.82</td>\n",
       "      <td>100.00</td>\n",
       "      <td>99.279087</td>\n",
       "      <td>66.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21985</th>\n",
       "      <td>s__Staphylococcus_haemolyticus--RS_GCF_0060943...</td>\n",
       "      <td>99.534795</td>\n",
       "      <td>99.62</td>\n",
       "      <td>97.61</td>\n",
       "      <td>99.81</td>\n",
       "      <td>106.122807</td>\n",
       "      <td>84.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22042</th>\n",
       "      <td>s__Stenotrophomonas_maltophilia--RS_GCF_900186...</td>\n",
       "      <td>99.628284</td>\n",
       "      <td>100.00</td>\n",
       "      <td>90.28</td>\n",
       "      <td>100.00</td>\n",
       "      <td>187.224422</td>\n",
       "      <td>115.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22134</th>\n",
       "      <td>s__Streptococcus_agalactiae--RS_GCF_000186445.1</td>\n",
       "      <td>99.804082</td>\n",
       "      <td>100.00</td>\n",
       "      <td>75.52</td>\n",
       "      <td>100.00</td>\n",
       "      <td>59.872898</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22256</th>\n",
       "      <td>s__Streptococcus_pneumoniae--RS_GCF_001457635.1</td>\n",
       "      <td>99.519055</td>\n",
       "      <td>99.60</td>\n",
       "      <td>91.29</td>\n",
       "      <td>100.00</td>\n",
       "      <td>84.233223</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22267</th>\n",
       "      <td>s__Streptococcus_pyogenes--RS_GCF_002055535.1</td>\n",
       "      <td>99.854260</td>\n",
       "      <td>99.85</td>\n",
       "      <td>97.16</td>\n",
       "      <td>100.00</td>\n",
       "      <td>20.476465</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "      <td>643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22327</th>\n",
       "      <td>s__Streptococcus_suis--RS_GCF_000294495.1</td>\n",
       "      <td>99.928587</td>\n",
       "      <td>100.00</td>\n",
       "      <td>93.29</td>\n",
       "      <td>100.00</td>\n",
       "      <td>70.876619</td>\n",
       "      <td>56.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27054</th>\n",
       "      <td>s__Vibrio_cholerae--RS_GCF_000621645.1</td>\n",
       "      <td>99.769454</td>\n",
       "      <td>99.98</td>\n",
       "      <td>88.12</td>\n",
       "      <td>100.00</td>\n",
       "      <td>143.633290</td>\n",
       "      <td>103.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27111</th>\n",
       "      <td>s__Vibrio_parahaemolyticus--RS_GCF_900460535.1</td>\n",
       "      <td>99.871602</td>\n",
       "      <td>100.00</td>\n",
       "      <td>88.60</td>\n",
       "      <td>100.00</td>\n",
       "      <td>225.351441</td>\n",
       "      <td>107.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27425</th>\n",
       "      <td>s__Wolbachia_pipientis--RS_GCF_021378375.1</td>\n",
       "      <td>99.011886</td>\n",
       "      <td>99.79</td>\n",
       "      <td>81.27</td>\n",
       "      <td>99.79</td>\n",
       "      <td>145.495536</td>\n",
       "      <td>126.5</td>\n",
       "      <td>1</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27554</th>\n",
       "      <td>s__Xanthomonas_oryzae--RS_GCF_004136375.1</td>\n",
       "      <td>99.364773</td>\n",
       "      <td>100.00</td>\n",
       "      <td>91.18</td>\n",
       "      <td>100.00</td>\n",
       "      <td>392.578283</td>\n",
       "      <td>450.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27618</th>\n",
       "      <td>s__Yersinia_pestis--RS_GCF_900460465.1</td>\n",
       "      <td>99.607197</td>\n",
       "      <td>99.79</td>\n",
       "      <td>94.49</td>\n",
       "      <td>99.97</td>\n",
       "      <td>359.932847</td>\n",
       "      <td>214.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1897</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              clade_name  checkm_average  \\\n",
       "765       s__Acinetobacter_baumannii--RS_GCF_009759685.1       99.821837   \n",
       "2461     s__Bacillus_A_bombysepticus--RS_GCF_006384875.1       98.933842   \n",
       "2551            s__Bacillus_subtilis--RS_GCF_000009045.1       99.592218   \n",
       "2556          s__Bacillus_velezensis--RS_GCF_001461825.1       99.618485   \n",
       "2778       s__Bifidobacterium_longum--RS_GCF_000196555.1       99.271630   \n",
       "3064         s__Bordetella_pertussis--RS_GCF_000306945.1       99.534642   \n",
       "3342          s__Brucella_melitensis--RS_GCF_000007125.1       99.095995   \n",
       "3409     s__Burkholderia_cenocepacia--RS_GCF_900446215.1       99.225651   \n",
       "3423          s__Burkholderia_mallei--RS_GCF_000011705.1       99.859408   \n",
       "3426     s__Burkholderia_multivorans--RS_GCF_000959525.1       99.853292   \n",
       "3450       s__Burkholderia_ubonensis--RS_GCF_902833085.1       99.815180   \n",
       "5020         s__Campylobacter_D_coli--RS_GCF_000254135.1       99.334832   \n",
       "5026       s__Campylobacter_D_jejuni--RS_GCF_001457695.1       99.234280   \n",
       "5506         s__Citrobacter_freundii--RS_GCF_011064845.1       99.772334   \n",
       "5535     s__Clostridioides_difficile--RS_GCF_001077535.1       98.996164   \n",
       "6223        s__Cronobacter_sakazakii--RS_GCF_000982825.1       99.680824   \n",
       "7409    s__Enterobacter_hormaechei_A--RS_GCF_001729745.1       99.533322   \n",
       "7456       s__Enterococcus_B_faecium--RS_GCF_001544255.1       99.570158   \n",
       "7491        s__Enterococcus_faecalis--RS_GCF_000392875.1       99.528996   \n",
       "7647             s__Escherichia_coli--RS_GCF_003697165.2       99.716624   \n",
       "8556       s__Francisella_tularensis--RS_GCF_000008985.1       99.700458   \n",
       "9746       s__Haemophilus_influenzae--RS_GCF_001457655.1       99.650459   \n",
       "10062         s__Helicobacter_pylori--RS_GCF_900478295.1       99.388257   \n",
       "10065       s__Helicobacter_pylori_C--RS_GCF_000277405.1       99.395205   \n",
       "11881    s__Klebsiella_michiganensis--RS_GCF_002925905.1       99.754060   \n",
       "11885       s__Klebsiella_pneumoniae--RS_GCF_000742135.1       99.859633   \n",
       "11886  s__Klebsiella_quasipneumoniae--RS_GCF_020525925.1       99.853109   \n",
       "11890        s__Klebsiella_variicola--RS_GCF_000828055.2       99.841411   \n",
       "12093  s__Lacticaseibacillus_rhamnosus--RS_GCF_900636...       99.234896   \n",
       "12107  s__Lactiplantibacillus_plantarum--RS_GCF_01413...       99.145888   \n",
       "12280      s__Legionella_pneumophila--RS_GCF_000008485.1       99.908789   \n",
       "12418      s__Leptospira_interrogans--RS_GCF_900156205.1       96.455399   \n",
       "12763      s__Listeria_monocytogenes--RS_GCF_900187225.1       99.410407   \n",
       "12764    s__Listeria_monocytogenes_B--RS_GCF_000307025.1       99.429626   \n",
       "14627     s__Mycobacterium_abscessus--RS_GCF_000069185.1       99.899507   \n",
       "14808  s__Mycobacterium_tuberculosis--RS_GCF_000195955.2       99.784281   \n",
       "15027       s__Neisseria_gonorrhoeae--RS_GCF_003315235.1       99.702851   \n",
       "15030      s__Neisseria_meningitidis--RS_GCF_900638555.1       99.631975   \n",
       "17272    s__Piscirickettsia_salmonis--RS_GCF_000297215.2       98.788795   \n",
       "18050           s__Proteus_mirabilis--RS_GCF_000160755.1       99.820749   \n",
       "18299     s__Pseudomonas_E_avellanae--RS_GCF_000444135.1       99.734828   \n",
       "18468    s__Pseudomonas_E_paracarnis--RS_GCF_904063055.1       99.210000   \n",
       "18663   s__Pseudomonas_E_viridiflava--RS_GCF_001642795.1       99.820437   \n",
       "18688      s__Pseudomonas_aeruginosa--RS_GCF_001457615.1       99.460522   \n",
       "21210         s__Salmonella_enterica--RS_GCF_000006945.2       99.922754   \n",
       "21964       s__Staphylococcus_aureus--RS_GCF_001027105.1       99.414489   \n",
       "21980  s__Staphylococcus_epidermidis--RS_GCF_006742205.1       99.604722   \n",
       "21985  s__Staphylococcus_haemolyticus--RS_GCF_0060943...       99.534795   \n",
       "22042  s__Stenotrophomonas_maltophilia--RS_GCF_900186...       99.628284   \n",
       "22134    s__Streptococcus_agalactiae--RS_GCF_000186445.1       99.804082   \n",
       "22256    s__Streptococcus_pneumoniae--RS_GCF_001457635.1       99.519055   \n",
       "22267      s__Streptococcus_pyogenes--RS_GCF_002055535.1       99.854260   \n",
       "22327          s__Streptococcus_suis--RS_GCF_000294495.1       99.928587   \n",
       "27054             s__Vibrio_cholerae--RS_GCF_000621645.1       99.769454   \n",
       "27111     s__Vibrio_parahaemolyticus--RS_GCF_900460535.1       99.871602   \n",
       "27425         s__Wolbachia_pipientis--RS_GCF_021378375.1       99.011886   \n",
       "27554          s__Xanthomonas_oryzae--RS_GCF_004136375.1       99.364773   \n",
       "27618             s__Yersinia_pestis--RS_GCF_900460465.1       99.607197   \n",
       "\n",
       "       checkm_median  checkm_min  checkm_max  contig_average  contig_median  \\\n",
       "765            99.98       57.57      100.00      138.579058           97.0   \n",
       "2461           99.18       54.46       99.69      183.734727          123.5   \n",
       "2551           99.81       74.14      100.00       43.341682           14.0   \n",
       "2556           99.67       86.05      100.00       36.077280            4.0   \n",
       "2778          100.00       66.59      100.00       65.646296           45.5   \n",
       "3064          100.00       93.11      100.00      194.961727            1.0   \n",
       "3342           99.45       87.85      100.00       32.812796           25.0   \n",
       "3409           99.73       92.74       99.87      160.485207          138.0   \n",
       "3423          100.00       95.25      100.00      128.373349          104.0   \n",
       "3426          100.00       95.95      100.00      103.606436           83.5   \n",
       "3450          100.00       96.13      100.00      130.868852          111.0   \n",
       "5020           99.44       94.89       99.72       58.293805           38.0   \n",
       "5026           99.53       85.63       99.88       56.876783           34.0   \n",
       "5506           99.90       83.75      100.00      127.188153           89.5   \n",
       "5535           99.45       77.59      100.00      290.946621          103.0   \n",
       "6223           99.93       90.86       99.99      129.543430           59.0   \n",
       "7409           99.67       57.05      100.00      107.831635           75.0   \n",
       "7456           99.63       88.42      100.00      168.013818          176.0   \n",
       "7491           99.63       75.03      100.00       65.322667           48.0   \n",
       "7647           99.93       55.50      100.00      197.466824          151.0   \n",
       "8556           99.82       93.88       99.93       96.071596           99.0   \n",
       "9746           99.67       98.35      100.00       55.768959           45.0   \n",
       "10062          99.42       93.07       99.77       51.165394           43.0   \n",
       "10065          99.42       88.05      100.00       28.668309           27.0   \n",
       "11881         100.00       74.64      100.00      140.378446          108.0   \n",
       "11885         100.00       70.27      100.00      126.444944          100.0   \n",
       "11886          99.99       93.58      100.00      127.797753           72.0   \n",
       "11890         100.00       85.89      100.00      111.490798           64.5   \n",
       "12093          99.46       91.01      100.00       89.692946           72.0   \n",
       "12107          99.38       90.87       99.38       88.426230           58.0   \n",
       "12280         100.00       93.89      100.00       53.524941           38.0   \n",
       "12418          96.47       92.94       97.65      229.484848          216.0   \n",
       "12763          99.45       90.66      100.00       35.770796           23.0   \n",
       "12764          99.45       93.99      100.00       33.034321           23.0   \n",
       "14627         100.00       94.00      100.00       39.924932           28.0   \n",
       "14808          99.94       93.00       99.94      102.057946           77.0   \n",
       "15027          99.82       97.31       99.89      116.353249           94.0   \n",
       "15030          99.75       95.48       99.83      155.097124          134.0   \n",
       "17272          98.81       98.24       98.85       34.614458            5.0   \n",
       "18050         100.00       75.44      100.00      161.011516           89.0   \n",
       "18299         100.00       91.01      100.00      327.241379          320.0   \n",
       "18468          99.80       78.79       99.80      154.647727          100.5   \n",
       "18663         100.00       96.43      100.00       90.516710           60.0   \n",
       "18688          99.68       81.03      100.00      163.096302          100.0   \n",
       "21210         100.00       58.97      100.00       93.888353           67.0   \n",
       "21964          99.51       83.21       99.70       61.840080           45.0   \n",
       "21980          99.81       59.82      100.00       99.279087           66.0   \n",
       "21985          99.62       97.61       99.81      106.122807           84.0   \n",
       "22042         100.00       90.28      100.00      187.224422          115.0   \n",
       "22134         100.00       75.52      100.00       59.872898           40.0   \n",
       "22256          99.60       91.29      100.00       84.233223           61.0   \n",
       "22267          99.85       97.16      100.00       20.476465           17.0   \n",
       "22327         100.00       93.29      100.00       70.876619           56.0   \n",
       "27054          99.98       88.12      100.00      143.633290          103.5   \n",
       "27111         100.00       88.60      100.00      225.351441          107.0   \n",
       "27425          99.79       81.27       99.79      145.495536          126.5   \n",
       "27554         100.00       91.18      100.00      392.578283          450.0   \n",
       "27618          99.79       94.49       99.97      359.932847          214.0   \n",
       "\n",
       "       contig_min  contig_max  \n",
       "765             1        1917  \n",
       "2461            1        1634  \n",
       "2551            1        1218  \n",
       "2556            1        1898  \n",
       "2778            1         975  \n",
       "3064            1        1366  \n",
       "3342            1         722  \n",
       "3409            3        1996  \n",
       "3423            1        1141  \n",
       "3426            2        1721  \n",
       "3450            3        1143  \n",
       "5020            1         577  \n",
       "5026            1         514  \n",
       "5506            1        1995  \n",
       "5535            1        1989  \n",
       "6223            1        1499  \n",
       "7409            1        1942  \n",
       "7456            1        1584  \n",
       "7491            1         967  \n",
       "7647            1        1956  \n",
       "8556            1         958  \n",
       "9746            1        1129  \n",
       "10062           1         432  \n",
       "10065           1         449  \n",
       "11881           1        1928  \n",
       "11885           1        1974  \n",
       "11886           1        1990  \n",
       "11890           1        1964  \n",
       "12093           1         558  \n",
       "12107           1         757  \n",
       "12280           1        1022  \n",
       "12418           2        1039  \n",
       "12763           1         935  \n",
       "12764           1         891  \n",
       "14627           1        1889  \n",
       "14808           1        1980  \n",
       "15027           1         882  \n",
       "15030           1         719  \n",
       "17272           2         547  \n",
       "18050           1        1700  \n",
       "18299           1        1765  \n",
       "18468           1        1931  \n",
       "18663           1        1741  \n",
       "18688           1        1996  \n",
       "21210           1        1979  \n",
       "21964           1        1654  \n",
       "21980           1        1725  \n",
       "21985           1        1086  \n",
       "22042           1        1605  \n",
       "22134           1         635  \n",
       "22256           1        1910  \n",
       "22267           1         643  \n",
       "22327           1        1219  \n",
       "27054           1        1806  \n",
       "27111           2        1868  \n",
       "27425           1         338  \n",
       "27554           1        1338  \n",
       "27618           1        1897  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/global/cfs/cdirs/kbase/ke_prototype/mcashman/analysis/dataframes/filtering_stage3/genome_list_A_SetOf58GPUclades.txt', 'r') as file:\n",
    "    string_list = [line.strip() for line in file]\n",
    "    \n",
    "grouped_df[grouped_df['clade_name'].isin(string_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a504ae9-5041-4385-92cd-35901c0a42dc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "clade = \"s__Streptomyces_griseoviridis_A--RS_GCF_005222485.1\"\n",
    "output_locations = ['/global/cfs/cdirs/kbase/ke_prototype/mcashman/motupan/motupan-batch/scripts/14_ANIthread99pt9999/out_sizeLT5','/global/cfs/cdirs/kbase/ke_prototype/mcashman/motupan/motupan-batch/scripts/14_ANIthread99pt9999/out_sizeLT500','/global/cfs/cdirs/kbase/ke_prototype/mcashman/motupan/motupan-batch/scripts/14_ANIthread99pt9999/out_sizeGTE500']\n",
    "\n",
    "out_file=pg.fetch_motupan_json_output_filepath(clade,output_locations)\n",
    "out_file\n",
    "\n",
    "df_temp = pd.read_json(out_file)\n",
    "df_temp = df_temp.T\n",
    "\n",
    "#df_temp.iloc[0]['aux_genome']\n",
    "#df_temp['gene_clusterss'][df_temp['gene_clusterss'].keys()[0]]['aa']\n",
    "df_temp['gene_clusterss'][df_temp['gene_clusterss'].keys()[0]]['genome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "90758dc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:40:43,963\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"2161914337\", \"msg\": \"Notebook started\"}\n",
      "{\"time\": \"2025-07-10 15:40:43,976\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"ConfigLoader initialized for target table: genome\"}\n",
      "{\"time\": \"2025-07-10 15:40:43,976\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"Resolved config path: s3a://cdm-lake/config-json/genome.json\"}\n",
      "{\"time\": \"2025-07-10 15:40:43,978\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"Loading config from MinIO: bucket=cdm-lake, key=config-json/genome.json\"}\n",
      "{\"time\": \"2025-07-10 15:40:44,065\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"Config loaded successfully from MinIO\"}\n",
      "{\"time\": \"2025-07-10 15:40:44,066\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"All required fields are present in the config\"}\n",
      "{\"time\": \"2025-07-10 15:40:44,066\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Target table: pangenome.genome\"}\n",
      "pangenome.genome\n",
      "{\"time\": \"2025-07-10 15:40:44,305\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"spark_session\", \"msg\": \"Found 4 JARs in /home/jovyan/jars\"}\n",
      "{\"time\": \"2025-07-10 15:40:44,306\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"spark_session\", \"msg\": \"JARs: ['/home/jovyan/jars/aws-java-sdk-bundle-1.12.367.jar', '/home/jovyan/jars/delta-core_2.12-2.4.0.jar', '/home/jovyan/jars/delta-storage-2.4.0.jar', '/home/jovyan/jars/hadoop-aws-3.3.4.jar']\"}\n",
      "{\"time\": \"2025-07-10 15:40:44,306\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"spark_session\", \"msg\": \"Starting Spark session...\"}\n",
      "{\"time\": \"2025-07-10 15:40:48,197\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"spark_session\", \"msg\": \"Spark session started successfully\"}\n"
     ]
    }
   ],
   "source": [
    "import boto3\n",
    "import yaml\n",
    "from src.logger import setup_logger\n",
    "from src.minio_uploader import upload_to_minio\n",
    "\n",
    "# Set up logger\n",
    "pipeline_name = \"pangenome\"\n",
    "target_table = \"genome\"\n",
    "schema = \"pangenome\"\n",
    "\n",
    "logger = setup_logger(\n",
    "    pipeline_name=pipeline_name,\n",
    "    target_table=target_table,\n",
    "    schema=schema\n",
    ")\n",
    "\n",
    "logger.info(\"Notebook started\")\n",
    "\n",
    "from config.config_loader import ConfigLoader\n",
    "\n",
    "# Path to config file in MinIO\n",
    "config_path = \"genome\"\n",
    "\n",
    "# Load config\n",
    "loader = ConfigLoader(config_path, logger)\n",
    "loader.load_and_validate()\n",
    "\n",
    "# Start Spark\n",
    "from src.spark_session import start_spark_session\n",
    "spark = start_spark_session(logger=logger)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "401e0c0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:40:51,040\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Input files: [{'source': 'GTDB', 'file_path': 's3a://cdm-lake/bronze/gtdb/sp_clusters_r214.tsv', 'file_type': 'tsv', 'delimiter': '\\t', 'ignore_first_line': 'no'}, {'source': 'NCBI', 'file_path': 's3a://cdm-lake/bronze/ncbi/assembly_summary_genbank.txt', 'file_type': 'tsv', 'delimiter': '\\t', 'ignore_first_line': 'yes'}]\"}\n",
      "{\"time\": \"2025-07-10 15:40:51,041\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"input_file_validator\", \"msg\": \"Starting input file validation...\"}\n",
      "Validating input files...\n",
      "\n",
      "{\"time\": \"2025-07-10 15:40:51,046\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"input_file_validator\", \"msg\": \"Checking file: s3a://cdm-lake/bronze/gtdb/sp_clusters_r214.tsv\"}\n",
      "🔍 Checking: s3a://cdm-lake/bronze/gtdb/sp_clusters_r214.tsv\n",
      "{\"time\": \"2025-07-10 15:40:51,053\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"input_file_validator\", \"msg\": \"File exists: s3://cdm-lake/bronze/gtdb/sp_clusters_r214.tsv\"}\n",
      "✅ File exists\n",
      "{\"time\": \"2025-07-10 15:40:51,179\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 1: Tab-delimited\"}\n",
      "   ✅ Line 1: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:40:51,180\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 2: Tab-delimited\"}\n",
      "   ✅ Line 2: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:40:51,181\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 3: Tab-delimited\"}\n",
      "   ✅ Line 3: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:40:51,182\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 4: Tab-delimited\"}\n",
      "   ✅ Line 4: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:40:51,182\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 5: Tab-delimited\"}\n",
      "   ✅ Line 5: Tab-delimited\n",
      "\n",
      "{\"time\": \"2025-07-10 15:40:51,183\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"input_file_validator\", \"msg\": \"Checking file: s3a://cdm-lake/bronze/ncbi/assembly_summary_genbank.txt\"}\n",
      "🔍 Checking: s3a://cdm-lake/bronze/ncbi/assembly_summary_genbank.txt\n",
      "{\"time\": \"2025-07-10 15:40:51,188\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"input_file_validator\", \"msg\": \"File exists: s3://cdm-lake/bronze/ncbi/assembly_summary_genbank.txt\"}\n",
      "✅ File exists\n",
      "{\"time\": \"2025-07-10 15:41:41,108\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 2: Tab-delimited\"}\n",
      "   ✅ Line 2: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:41:41,207\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 3: Tab-delimited\"}\n",
      "   ✅ Line 3: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:41:41,213\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 4: Tab-delimited\"}\n",
      "   ✅ Line 4: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:41:41,214\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 5: Tab-delimited\"}\n",
      "   ✅ Line 5: Tab-delimited\n",
      "{\"time\": \"2025-07-10 15:41:41,216\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"input_file_validator\", \"msg\": \"Line 6: Tab-delimited\"}\n",
      "   ✅ Line 6: Tab-delimited\n",
      "\n",
      "{\"time\": \"2025-07-10 15:41:51,368\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Starting schema vs. file validation\"}\n",
      "{\"time\": \"2025-07-10 15:41:51,522\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Output file: {'output_file_path': 's3a://cdm-lake/pangenome-source/table_genome_V1.1.tsv', 'file_type': 'tsv', 'delimiter': '\\t', 'ignore_first_line': 'no'}\"}\n",
      "{\"time\": \"2025-07-10 15:41:51,525\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Output file: s3a://cdm-lake/pangenome-source/table_genome_V1.1.tsv\"}\n",
      "{\"time\": \"2025-07-10 15:41:51,526\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"validate_schema\", \"msg\": \"Delimiter: \t, Ignore first line: False\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,400\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Header columns successfully extracted from output file\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,402\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"validate_schema\", \"msg\": \"Header columns: ['genome_id', 'gtdb_species_clade_id', 'gtdb_taxonomy_id', 'ncbi_biosample_id', 'fna_file_path_nersc', 'faa_file_path_nersc']\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,402\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Schema file path: s3a://cdm-lake/schemas/pangenome-schema.yaml\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,403\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Schema file: s3a://cdm-lake/schemas/pangenome-schema.yaml\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,470\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Target table: pangenome.genome\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,471\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Derived schema class name: genome\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,473\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Schema loaded and class found successfully\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,473\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"validate_schema\", \"msg\": \"Schema columns: ['genome_id', 'gtdb_taxonomy_id', 'gtdb_species_clade_id', 'ncbi_biosample_id', 'fna_file_path_nersc', 'faa_file_path_nersc']\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,474\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"Comparison between schema and file header completed\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,475\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"All schema columns are present in the file\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,475\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"validate_schema\", \"msg\": \"No extra columns in the file\"}\n",
      "🔹 Header columns from file:\n",
      "['genome_id', 'gtdb_species_clade_id', 'gtdb_taxonomy_id', 'ncbi_biosample_id', 'fna_file_path_nersc', 'faa_file_path_nersc']\n",
      "\n",
      "✅ Columns in schema class 'genome':\n",
      "['genome_id', 'gtdb_taxonomy_id', 'gtdb_species_clade_id', 'ncbi_biosample_id', 'fna_file_path_nersc', 'faa_file_path_nersc']\n",
      "\n",
      "🔍 Comparison results:\n",
      "✅ All schema columns are present in file\n",
      "✅ No extra columns in file\n",
      "{\"time\": \"2025-07-10 15:41:52,495\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Output file: {'output_file_path': 's3a://cdm-lake/pangenome-source/table_genome_V1.1.tsv', 'file_type': 'tsv', 'delimiter': '\\t', 'ignore_first_line': 'no'}\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,496\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Starting validation on file: s3a://cdm-lake/pangenome-source/table_genome_V1.1.tsv\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:41:52,497\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_validations\", \"msg\": \"Delimiter: \t, Ignore first line: False\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,498\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Validations: [{'column': 'genome_id', 'validation_type': 'not_null', 'error_message': 'Missing genome_id'}, {'column': 'genome_id', 'validation_type': 'regex_match', 'pattern': '^[A-Za-z0-9_.-]+$', 'error_message': 'Invalid genome_id format'}, {'column': 'gtdb_taxonomy_id', 'validation_type': 'not_null', 'error_message': 'Missing taxonomy ID'}, {'column': 'gtdb_species_clade_id', 'validation_type': 'not_null', 'error_message': 'Missing species clade ID'}]\"}\n",
      "{\"time\": \"2025-07-10 15:41:52,498\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_validations\", \"msg\": \"Validation rules: [{'column': 'genome_id', 'validation_type': 'not_null', 'error_message': 'Missing genome_id'}, {'column': 'genome_id', 'validation_type': 'regex_match', 'pattern': '^[A-Za-z0-9_.-]+$', 'error_message': 'Invalid genome_id format'}, {'column': 'gtdb_taxonomy_id', 'validation_type': 'not_null', 'error_message': 'Missing taxonomy ID'}, {'column': 'gtdb_species_clade_id', 'validation_type': 'not_null', 'error_message': 'Missing species clade ID'}]\"}\n",
      "\n",
      "🔍 Running validations:\n",
      "\n",
      "{\"time\": \"2025-07-10 15:41:55,567\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Running validations...\"}\n",
      "{\"time\": \"2025-07-10 15:41:55,569\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_validations\", \"msg\": \"Header columns: ['genome_id', 'gtdb_species_clade_id', 'gtdb_taxonomy_id', 'ncbi_biosample_id', 'fna_file_path_nersc', 'faa_file_path_nersc']\"}\n",
      "\n",
      "📋 Validation Summary:\n",
      "{\"time\": \"2025-07-10 15:41:56,406\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Validation summary:\"}\n",
      "✅ All validation rules ran successfully. No issues found.\n",
      "{\"time\": \"2025-07-10 15:41:56,410\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"All validation rules passed. No issues found.\"}\n",
      "\n",
      "✅ Passed Rules:\n",
      "✔️ genome_id - not_null\n",
      "{\"time\": \"2025-07-10 15:41:56,412\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Passed: genome_id - not_null\"}\n",
      "✔️ genome_id - regex_match\n",
      "{\"time\": \"2025-07-10 15:41:56,414\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Passed: genome_id - regex_match\"}\n",
      "✔️ gtdb_species_clade_id - not_null\n",
      "{\"time\": \"2025-07-10 15:41:56,420\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Passed: gtdb_species_clade_id - not_null\"}\n",
      "✔️ gtdb_taxonomy_id - not_null\n",
      "{\"time\": \"2025-07-10 15:41:56,422\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_validations\", \"msg\": \"Passed: gtdb_taxonomy_id - not_null\"}\n",
      "{\"time\": \"2025-07-10 15:41:56,540\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"Loading config from MinIO: bucket=cdm-lake, key=config-json/genome.json\"}\n",
      "{\"time\": \"2025-07-10 15:41:56,554\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"Config loaded successfully from MinIO\"}\n",
      "{\"time\": \"2025-07-10 15:41:56,556\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"config_loader\", \"msg\": \"All required fields are present in the config\"}\n",
      "{\"time\": \"2025-07-10 15:41:56,556\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Referential integrity rules: [{'foreign_key': 'genome_id', 'reference_table': 'gtdb_metadata', 'reference_column': 'accession', 'database': 'pangenome', 'action': 'log'}, {'foreign_key': 'gtdb_species_clade_id', 'reference_table': 'gtdb_species_clade', 'reference_column': 'gtdb_species_clade_id', 'database': 'pangenome', 'action': 'log'}, {'foreign_key': 'gtdb_taxonomy_id', 'reference_table': 'gtdb_taxonomy_r214v1', 'reference_column': 'gtdb_taxonomy_id', 'database': 'pangenome', 'action': 'log'}]\"}\n",
      "{\"time\": \"2025-07-10 15:41:56,557\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Target table: pangenome.genome\"}\n",
      "{\"time\": \"2025-07-10 15:42:04,544\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Starting referential integrity checks for table: pangenome.genome\"}\n",
      "{\"time\": \"2025-07-10 15:42:04,545\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Checking foreign key: genome_id in pangenome.genome → accession in pangenome.gtdb_metadata\"}\n",
      "\n",
      "🔍 Checking FK `genome_id` in `pangenome.genome` → `accession` in `pangenome.gtdb_metadata`\n",
      "{\"time\": \"2025-07-10 15:42:10,950\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Violations found: 0\"}\n",
      " → Violations found: 0\n",
      "{\"time\": \"2025-07-10 15:42:10,951\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Checking foreign key: gtdb_species_clade_id in pangenome.genome → gtdb_species_clade_id in pangenome.gtdb_species_clade\"}\n",
      "\n",
      "🔍 Checking FK `gtdb_species_clade_id` in `pangenome.genome` → `gtdb_species_clade_id` in `pangenome.gtdb_species_clade`\n",
      "{\"time\": \"2025-07-10 15:42:12,351\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Violations found: 0\"}\n",
      " → Violations found: 0\n",
      "{\"time\": \"2025-07-10 15:42:12,352\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Checking foreign key: gtdb_taxonomy_id in pangenome.genome → gtdb_taxonomy_id in pangenome.gtdb_taxonomy_r214v1\"}\n",
      "\n",
      "🔍 Checking FK `gtdb_taxonomy_id` in `pangenome.genome` → `gtdb_taxonomy_id` in `pangenome.gtdb_taxonomy_r214v1`\n",
      "{\"time\": \"2025-07-10 15:42:13,865\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Violations found: 293059\"}\n",
      " → Violations found: 293059\n",
      " → Sample violations:\n",
      "{\"time\": \"2025-07-10 15:42:13,867\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"WARNING\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"293059 violations found in referential integrity check for FK: gtdb_taxonomy_id → gtdb_taxonomy_id in pangenome.gtdb_taxonomy_r214v1\"}\n",
      "{\"time\": \"2025-07-10 15:42:15,018\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"ERROR\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Sample violations for FK 'gtdb_taxonomy_id': [{'gtdb_taxonomy_id': 'd__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas_A'}, {'gtdb_taxonomy_id': 'd__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas_A'}, {'gtdb_taxonomy_id': 'd__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas_A'}, {'gtdb_taxonomy_id': 'd__Bacteria;p__Chloroflexota;c__Dehalococcoidia;o__Bin125;f__Bin125;g__Bin125'}, {'gtdb_taxonomy_id': 'd__Bacteria;p__Chloroflexota;c__Dehalococcoidia;o__Bin125;f__Bin125;g__Bin125'}]\"}\n",
      "{\"time\": \"2025-07-10 15:42:15,024\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Logging config: {'error_table': 'genome_errors', 'output_delta_path': 's3a://cdm-lake/logs/errors/genome'}\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:19,392\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"ERROR\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Referential integrity violations written to Delta table: pangenome.genome_errors at s3a://cdm-lake/logs/errors/genome\"}\n",
      "\n",
      "🚨 All violations logged to: pangenome.genome_errors\n",
      "{\"time\": \"2025-07-10 15:42:19,394\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"WARNING\", \"module\": \"run_referential_integrity_checks\", \"msg\": \"Referential integrity check(s) failed.\"}\n",
      "\n",
      "❌ Some referential integrity checks failed.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Validate input files\n",
    "from src.input_file_validator import validate_input_files\n",
    "validate_input_files(loader, logger)\n",
    "\n",
    "\n",
    "# Check schema headers\n",
    "from src.validate_schema import validate_schema_against_file\n",
    "validate_schema_against_file(loader, logger)\n",
    "\n",
    "\n",
    "# Run validations\n",
    "from src.run_validations import run_validations_from_config\n",
    "run_validations_from_config(loader, logger)\n",
    "\n",
    "\n",
    "# Referential integrity\n",
    "from src.run_referential_integrity_checks import run_referential_integrity_checks\n",
    "run_referential_integrity_checks(spark, loader, logger)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f2bfb80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:20,910\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"config_loader\", \"msg\": \"Target table: pangenome.genome\"}\n",
      "{\"time\": \"2025-07-10 15:42:20,911\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Starting Great Expectations validation for table: pangenome.genome with suite: default_suite\"}\n",
      "{\"time\": \"2025-07-10 15:42:20,965\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Loaded Spark table: pangenome.genome\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,126\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Great Expectations context initialized.\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,151\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Expectation suite 'default_suite' added or updated.\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,237\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Spark datasource registered in GE context.\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,238\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Batch request created for data asset: pangenome_genome\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,395\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Validator initialized for batch request.\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,396\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Found 6 expectations to apply.\"}\n",
      "{\"time\": \"2025-07-10 15:42:21,396\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Loaded GE validations config: [{'expectation_type': 'expect_column_values_to_not_be_null', 'params': {'column': 'genome_id'}}, {'expectation_type': 'expect_column_values_to_match_regex', 'params': {'column': 'genome_id', 'regex': '^[A-Za-z0-9_.-]+$'}}, {'expectation_type': 'expect_column_values_to_not_be_null', 'params': {'column': 'gtdb_species_clade_id'}}, {'expectation_type': 'expect_column_values_to_not_be_null', 'params': {'column': 'gtdb_taxonomy_id'}}, {'expectation_type': 'expect_column_to_exist', 'params': {'column': 'ncbi_biosample_id'}}, {'expectation_type': 'expect_column_to_exist', 'params': {'column': 'sample_id'}}]\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3097c56c0aaa49a5a6ee06223f176b07",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:22,279\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_values_to_not_be_null with args {'column': 'genome_id'} → result: True\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dbbefaf670f64634aba09a2ba51e51a0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/11 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:22,639\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_values_to_match_regex with args {'column': 'genome_id', 'regex': '^[A-Za-z0-9_.-]+$'} → result: True\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "03bc039e1db6449db5f9d1e3dc0d81c4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:22,853\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_values_to_not_be_null with args {'column': 'gtdb_species_clade_id'} → result: True\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5518fb4872464e58a1d95f16d7733a1b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:23,057\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_values_to_not_be_null with args {'column': 'gtdb_taxonomy_id'} → result: True\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7d2fb7c8e40f473187de431e5d6b57c8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:23,072\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_to_exist with args {'column': 'ncbi_biosample_id'} → result: True\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f8546ab412ae41909e914e20b2d40bfe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:23,086\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Applied GE expectation: expect_column_to_exist with args {'column': 'sample_id'} → result: False\"}\n",
      "{\"time\": \"2025-07-10 15:42:23,119\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Expectations saved to suite.\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be14ee95401042698cc2e417d46a0aa8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:23,667\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Validation run completed.\"}\n",
      "{\"time\": \"2025-07-10 15:42:23,672\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"DEBUG\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Validation summary: {'success': False, 'successful_expectations': 5, 'unsuccessful_expectations': 1, 'success_percent': 83.33333333333334}\"}\n",
      "{\"time\": \"2025-07-10 15:42:25,809\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Data Docs built after validation.\"}\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1027abe4747d405b9aceece1e3de207b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Calculating Metrics:   0%|          | 0/28 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"time\": \"2025-07-10 15:42:27,578\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Checkpoint executed.\"}\n",
      "{\"time\": \"2025-07-10 15:42:29,541\", \"pipeline\": \"pangenome\", \"schema\": \"pangenome\", \"table\": \"genome\", \"level\": \"INFO\", \"module\": \"run_great_expectations_validations\", \"msg\": \"Data Docs rebuilt with checkpoint results.\"}\n",
      "✅ GE validation and checkpoint complete. Data Docs generated.\n",
      "✅ Uploaded log to MinIO at: s3://cdm-lake/logs/pangenome/pipeline_run_20250710_154043.log\n"
     ]
    }
   ],
   "source": [
    "# Great Expectations\n",
    "from src.run_great_expectations_validations import run_great_expectations_validation\n",
    "run_great_expectations_validation(spark, loader, logger)\n",
    "\n",
    "\n",
    "# Upload log file to MinIO\n",
    "upload_to_minio(logger.log_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7806b2be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
